<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/library_prep/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Manual CellRanger library preparation - scRNAbox</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Manual CellRanger library preparation";
        var mkdocs_page_input_path = "library_prep.md";
        var mkdocs_page_url = "/library_prep/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scRNAbox
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipeline</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 0: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 1: FASTQ to expression matrix</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step2/">Step 2: Seurat object and ambient RNA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step3/">Step 3: Quality control and filtering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4/">Step 4: Doublet removal (standard)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4HTO/">Step 4: Demultiplexing (HTO)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step5/">Step 5: Integration</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step6/">Step 6: Clustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step7/">Step 7: Cluster annotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step8/">Step 8: Differential gene expression</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Additional information</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../config/">Job configurations</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset1/">scRNAbox analysis of midbrain dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading the midbrain dataset</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Manual CellRanger library preparation</a>
    <ul class="current">
    </ul>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../FAQ/">FAQ</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">- License</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../changelog/">- Changelog</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">- Contributing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">- Acknowledgement</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scRNAbox</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Tutorial &raquo;</li>
      <li class="breadcrumb-item active">Manual CellRanger library preparation</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="cellranger-library-preparation">CellRanger Library Preparation</h2>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#standard-analysis-track">Standard Analysis Track</a><ul>
<li><a href="#librarycsv">library.csv</a></li>
</ul>
</li>
<li><a href="#cell-hashtag-analysis-track">Cell Hashtag Analysis Track</a><ul>
<li><a href="#librarycsv">library.csv</a></li>
<li><a href="#feature-ref-csv">feature_ref.csv</a></li>
</ul>
</li>
</ul>
<hr />
<h2 id="introduction">Introduction</h2>
<p>Step 1 (FASTQ to gene expression matrix) of the scRNAbox pipeline leverages the CellRanger <em>counts</em> pipeline to generate gene expression matrices from FASTQ files. In order to run the CellRanger <em>counts</em> pipeline, libraries must be generated to define the information of the FASTQ files. For the <a href="">Standard Analysis Track</a>, a <code>library.csv</code> file must be generated for each sample. For the <a href="">Cell Hashtag Analysis Track</a>, a <code>library.csv</code> and <code>feature_ref.csv</code> file must be generated for each sequencing run. Athough scRNAbox provides an option to automate the library preparation process, it is important that users understand the information that is required for these files. </p>
<p>In this tutorial we demonstrate how users can manually prepare the <code>library.csv</code> and <code>feature_ref.csv</code> files. The information presented in this tutorial can also be used to inform the parameters that must be defined by the user for automated library preparation.</p>
<hr />
<h2 id="standard-analysis-track">Standard Analysis Track</h2>
<h4 id="librarycsv">library.csv</h4>
<p>The <code>library.csv</code> file defines the necessary information of the FASTQ files for the experiment, including the gene expression and antibody assays. The structure of the <code>library.csv</code> file should be: <br /></p>
<pre><code>fastqs,sample,library_type
path/to/fastqs/directory/,SampleName,Gene Expression
</code></pre>
<ul>
<li>The <code>fastqs</code> column defines the path to the directory that contains the FASTQ files for the experiment. <br /> </li>
<li>The <code>sample</code> column defines the sample name of the corresponding FASTQ file. Please note that FASTQ files must be named according to standard CellRanger nomenclature (e.g. CTRL1_S1_L001_R1_001.fastq). For more information, please visit CellRanger's <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input">documentation</a>. <br /></li>
<li>The <code>library_type</code> column defines the assay type. For the Standard Analysis Track, the <code>library_type</code> will be "Gene Expression". <br /></li>
</ul>
<h4 id="example-manual-library-preparation">Example: manual library preparation</h4>
<p>For the Standard Analysis Track, if the experiment comprises four samples (two case and two controls), the following steps should be taken for manual library preparation: <br /></p>
<p>1) Navigate to the working directory and create a <code>samples_info</code> folder: <br /></p>
<pre><code>cd ~/working_directory
mkdir samples_info
</code></pre>
<p>2) Navigate to the <code>samples_info</code> folder and create a folder for each sample run: <br /></p>
<pre><code>cd samples_info
mkdir case1
mkdir case2
mkdir control1
mkdir control2
</code></pre>
<p>3) Navigate to the folder for each sample and create the <code>library.csv</code> file. <br /></p>
<p>After performing steps 1-3 above, the structure of the samples_info folder (<code>~working_directory/samples_info</code>) for an experiment with four samples should be:</p>
<pre><code>working_directory
└── samples_info
    ├── case1
    │   └──library.csv
    ├── case2
    │   └── library.csv
    ├── control1
    │   └── library.csv
    └── control2
        └── library.csv
</code></pre>
<hr />
<h2 id="cell-hashtag-analysis-track">Cell Hashtag Analysis Track</h2>
<h4 id="librarycsv_1">library.csv</h4>
<p>The <code>library.csv</code> file defines the necessary information of the FASTQ files for the experiment, including the gene expression and antibody assays. The structure of the <code>library.csv</code> file should be: <br /></p>
<pre><code>fastqs,sample,library_type
path/to/fastqs/directory/,SampleNameGEX,Gene Expression
path/to/fastqs/directory/,SampleNameHTO,Antibody Capture
</code></pre>
<ul>
<li>The <code>fastqs</code> column defines the path to the directory that contains the FASTQ files for the experiment. <br /> </li>
<li>The <code>sample</code> column defines the sample name of the corresponding FASTQ file. Please note that FASTQ files must be named according to standard CellRanger nomenclature (e.g. CTRL1_S1_L001_R1_001.fastq). For more information please visit CellRanger's <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input">documentation</a>. <br /></li>
<li>The <code>library_type</code> column defines the assay type. For the Cell Hashtag Analysis track, each sequencing run should have a "Gene Expression" and "Antibody Capture" assay. For more information, please visit CellRanger's <a href="" title="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/feature-bc-analysis">documentation</a> <br /></li>
</ul>
<h4 id="feature_refcsv">feature_ref.csv</h4>
<p>The <code>feature_ref.csv</code> file defines the necessary information for processing the sample-specific barcodes that will eventually be used to demultiplex the pooled samples. For example, if there are four samples pooled together with four unique barcode identifiers, the structure of the <code>feature_ref.csv</code> file should be:</p>
<pre><code>id,name,read,pattern,sequence,feature_type
Hash1,B0251_TotalSeqB,R2,5PNNNNNNNNNN(BC),GTCAACTCTTTAGCG,Antibody Capture
Hash2,B0252_TotalSeqB,R2,5PNNNNNNNNNN(BC),TGATGGCCTATTGGG,Antibody Capture
Hash3,B0253_TotalSeqB,R2,5PNNNNNNNNNN(BC),TTCCGCCTCTCTTTG,Antibody Capture
Hash4,B0254_TotalSeqB,R2,5PNNNNNNNNNN(BC),AGTAAGTTCAGCGTA,Antibody Capture
</code></pre>
<ul>
<li>The <code>id</code> column defines the barcode ID which will be used to track the feature counts. <br /> </li>
<li>The <code>name</code> column defines the arbitrary name for the barcode identifier. <br /> </li>
<li>The <code>read</code> column defines which RNA sequencing read contains the barcode sequence. This value Will be either R1 or R2.<br /> </li>
<li>The <code>pattern</code> column defines the pattern of the barcode identifiers. For more information please visit the 10X Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/feature-bc-analysis#pattern">documentation</a><br /> </li>
<li>The <code>sequence</code> column defines nucleotide sequence associated with the barcode identifier.<br /> </li>
<li>The <code>feature_type</code> column defines the type of feature used for sample identification. Please ensure that the feature_type in the <code>feature_ref.csv</code> file matches a library_type in the <code>library.csv</code> file.  <br /> </li>
</ul>
<p>For more information regarding the preparation of the <code>feature_ref.csv</code>, please see CellRanger's <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/feature-bc-analysis">documentation</a>.</p>
<h4 id="example-manual-library-preparation_1">Example: manual library preparation</h4>
<p>For the Cell Hashtag Analysis Track, if the experiment comprises four sequencing, the following steps should be taken for manual library preparation: <br /></p>
<p>1) Navigate to the working directory and create a <code>samples_info</code> folder: <br /></p>
<pre><code>cd ~/working_directory
mkdir samples_info
</code></pre>
<p>2) Navigate to the <code>samples_info</code> folder and create a folder for each sequencing run: <br /></p>
<pre><code>cd samples_info
mkdir run1
mkdir run2
mkdir run3
mkdir run4
</code></pre>
<p>3) Navigate to the folder for each sequencing and create the <code>library.csv</code> file. <br /></p>
<p>4) Navigate to the folder for each sequencing and create the <code>feature_ref.csv</code> file. <br /></p>
<p>After performing steps 1-4 above, the structure of the samples_info folder (<code>~working_directory/samples_info</code>) for an experiment with four sequencing runs should be:</p>
<pre><code>working_directory
├── samples_info
    ├── run1
    │    ├── library.csv
    │    └── feature_ref.csv
    ├── run2
    │    ├── library.csv
    │    └── feature_ref.csv
    ├── run3
    │    ├── library.csv
    │    └── feature_ref.csv   
    └── run4
        ├── library.csv
        └── feature_ref.csv   
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../midbrain_download/" class="btn btn-neutral float-left" title="Downloading the midbrain dataset"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../FAQ/" class="btn btn-neutral float-right" title="FAQ">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../midbrain_download/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../FAQ/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
