<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" /><link rel="canonical" href="https://neurobioinfo.github.io/" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>scrnabox</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Home";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = "/";
      </script>
    
    <script src="js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> scrnabox
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href=".">Home</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#contents">Contents</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#workflow-analysis">Workflow analysis</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#standard-scrna-seq">Standard scRNA-seq</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell-hashtags">Cell Hashtags</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#scrnaboxslurm">scrnabox.slurm</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#installing">Installing</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#processed-data">Processed Data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tutorial">Tutorial</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#references">References</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#contributing">Contributing</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#changelog">Changelog</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#license">License</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#acknowledgement">Acknowledgement</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#todo">Todo</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="README_SCRNA/">Standard scRNA</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="README_HTO/">Cell hashtags</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="about/">About</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">scrnabox</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li>Home</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="scrnabox-a-pipeline-for-scrna-analysis-under-hpc">scrnabox: A pipeline for scRNA analysis under HPC</h1>
<p>This repository includes a tutorial for the scRNA-seq data analysis using scrnabox pipeline under an HPC environment (<a href="https://slurm.schedmd.com/">slurm work load manager system</a>). </p>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#workflow-analysis">Workflow analysis</a></li>
<li><a href="#standard-scRNA-seq">Standard scRNA-seq</a></li>
<li><a href="#cell-hashtags">Cell hashtags</a></li>
<li><a href="#scrnaboxsvn">scrnabox.slurm</a></li>
<li><a href="#installing">Installing</a></li>
<li><a href="#processed-data">Processed Data</a></li>
<li><a href="#tutorial">Tutorial</a></li>
<li><a href="#references">References</a></li>
</ul>
<hr />
<h2 id="workflow-analysis">Workflow analysis</h2>
<p>The following figures illustrate the steps involved in analyzing scRNA-seq data using the Standard and Cell Hashtags with the scrnabox pipeline
<br />
<br />
<kbd>
<img alt="Steps of Standard scRNA-seq " src="https://raw.githubusercontent.com/neurobioinfo/scrnabox/main/figs/scrna.png" />
</kbd>
<kbd>
<img alt="Steps of Cell Hashtags scRNA-seq" src="https://raw.githubusercontent.com/neurobioinfo/scrnabox/main/figs/hto.png" />
</kbd></p>
<p>This pipeline currently includes implementation of the standard and cell hashtags scRNA.</p>
<h4 id="standard-scrna-seq"><a href="https://github.com/neurobioinfo/scrnabox/tree/main/README_SCRNA.md">Standard scRNA-seq</a></h4>
<p>The following steps describe how to analyze scRNA-seq data using the pipeline:<br />
- Step 1: cellranger - This step runs Cellranger on the scRNA-seq data to generate the feature-barcode matrices for each sample.<br />
- Step 2: Seurat object - This step runs Seurat on the feature-barcode matrices obtained from step 1 to generate a Seurat object for each sample. The Seurat object contains a lot of information, and accessing SeuratObject@meta.data will provide a data frame with relevant information on each cell.<br />
- Step 3: QC and filter - The Seurat object includes quality metrics that can be used to filter cells and genes against possible doublets. Metrics such as total UMI counts per cell (nCount_RNA), total number of detected features per cell (nFeature_RNA), and mitochondrial count (percent.mito) are often used.<br />
- Step 4: Doublet removal - This step can be used to remove doublets from the data. By default, the pipeline removes doublets, but you can choose to keep them by changing the parameter from 'yes' to 'no'.<br />
- Step 5: Integration - This step integrates multiple scRNA-seq datasets using the Comprehensive Integration of Single Cell Data (CCA) method in Seurat, Tim, et al. (2019). The pipeline identifies anchors using the FindIntegrationAnchors function and passes them to the IntegrateData function to get a single Seurat object representing all the datasets.<br />
- Step 6: Clustering- It involves clustering the data using a k-nearest neighbor graph based on the integrated PCA. This step produces UMAP and heatmaps of unlabelled clusters. However, it is up to the user to decide on the best cluster resolution outside the pipeline by examining the output and selecting the most appropriate annotation for the clusters.<br />
- step 7: Cluster annotate - In this step, you can annotate the clusters with known cell types or use marker genes to predict the cell type of each cluster. <br />
- step 8: Differential gene expression (DEG)- There are multiple ways to perform differential gene expression analysis, but in this pipeline, we use the FindAllMarkers function to rank the highly differentially expressed genes in each cluster, which allows us to identify genes that are significantly differentially expressed between each cluster and the rest of the cells. From there, we can define contrasts to run statistical tests and investigate the phenotype and genotypes of each cluster.<br /> <br />
The Step 1 - Step 8 can be done using <a href="https://github.com/neurobioinfo/scrnabox/tree/main/README_SCRNA.md">scrnabox.slurm</a> in the HPC system (<a href="https://slurm.schedmd.com/">slurm work load manager system</a>).
- step 9, Enrichment analysis: in this step, we obtain a list of significant genes using enrichment methods. The step 9 can be done using scrnaboxR, see <a href="https://github.com/neurobioinfo/scrnabox/blob/main/tutorial/practice.md">Practice</a>.</p>
<h4 id="cell-hashtags"><a href="https://github.com/neurobioinfo/scrnabox/tree/main/README_HTO.md">Cell Hashtags</a></h4>
<p>The following steps explain how to analyze the Hashtag oligonucleotide (henceforth referred to as HTO)
- Step 1: cellranger - This step runs Cellranger on the scRNA-seq data to generate the feature-barcode matrices for each sample.
- Step 2: Seurat object - This step runs Seurat on the feature-barcode matrices obtained from step 1 to generate a Seurat object for each sample. The Seurat object contains a lot of information, and accessing SeuratObject@meta.data will provide a data frame with relevant information on each cell.
- Step 3: QC and filter - The Seurat object includes quality metrics that can be used to filter cells and genes against possible doublets. Metrics such as total UMI counts per cell (nCount_RNA), total number of detected features per cell (nFeature_RNA), and mitochondrial count (percent.mito) are often used.
- Step 4: Demuplixing- In this step, demultiplexing is performed to separate the reads in the sequencing run according to their sample of origin, based on their barcode information. The pipeline includes an option to remove doublets and negative cells during a later step after quality control and filtering.
- Step 5: Integration - This step integrates multiple scRNA-seq datasets using the Comprehensive Integration of Single Cell Data (CCA) method in Seurat, Tim, et al. (2019). The pipeline identifies anchors using the FindIntegrationAnchors function and passes them to the IntegrateData function to get a single Seurat object representing all the datasets.
- Step 6: Clustering- It involves clustering the data using a k-nearest neighbor graph based on the integrated PCA. This step produces UMAP and heatmaps of unlabelled clusters. However, it is up to the user to decide on the best cluster resolution outside the pipeline by examining the output and selecting the most appropriate annotation for the clusters.
- step 7: Cluster annotate - In this step, you can annotate the clusters with known cell types or use marker genes to predict the cell type of each cluster. 
- step 8: Differential gene expression (DEG)- There are multiple ways to perform differential gene expression analysis, but in this pipeline, we use the FindAllMarkers function to rank the highly differentially expressed genes in each cluster, which allows us to identify genes that are significantly differentially expressed between each cluster and the rest of the cells. From there, we can define contrasts to run statistical tests and investigate the phenotype and genotypes of each cluster.</p>
<p>The Step 1 - Step 8 can be done using <a href="https://github.com/neurobioinfo/scrnabox/tree/main/README_HTO.md">scrnabox.slurm</a> in the HPC system (<a href="https://slurm.schedmd.com/">slurm work load manager system</a>).
- step 9, Enrichment analysis: in this step, we obtain a list of significant genes using enrichment methods. The step 9 can be done using scrnaboxR, see <a href="https://github.com/neurobioinfo/scrnabox/blob/main/tutorial/practice.md">Practice</a>.</p>
<h2 id="scrnaboxslurm"><a href="https://github.com/neurobioinfo/scrnabox/tree/main/scrnabox.slurm">scrnabox.slurm</a></h2>
<p>The <code>scrnabox.slurm</code> is a pipeline specifically designed for analyzing data under a High-Performance Computing (HPC) system  using the Slurm Workload Manager (<a href="https://slurm.schedmd.com/">slurm work load manager system</a>). 
It has been extensively utilized on the  <a href="https://docs.alliancecan.ca/wiki/B%C3%A9luga">Beluga</a>, you can find comprehensive instructions on how to utilize the pipeline 
for <a href="https://github.com/neurobioinfo/scrnabox/tree/main/README_HTO.md">Cell Hashtags</a> and <a href="https://github.com/neurobioinfo/scrnabox/tree/main/README_SCRNA.md">Standard scRNA-seq</a>, 
the pipeline provides detailed guidance on processing and analyzing these specific types of single-cell RNA sequencing (scRNA-seq) data. </p>
<h2 id="installing">Installing</h2>
<p>The package is written in the bash, so it can be used with any slurm system. To download 
<code>scrnabox.slurm</code> run the below comments </p>
<pre><code>wget https://github.com/neurobioinfo/scrnabox/releases/download/v0.1.0/scrnabox.slurm.zip
unzip scrnabox.slurm.zip 
</code></pre>
<p>To obtain a brief guidance of the pipeline, execute the following code.</p>
<pre><code>bash ./scrnabox.slurm/launch_scrnabox.sh -h 
</code></pre>
<p><code>scrnabox.slurm</code> needs <code>R</code> and <code>cellranger</code>. For the R, you need to install 
<code>'Seurat','ggplot2', 'dplyr', 'foreach', 'doParallel', 'Matrix', 'DoubletFinder','cowplot','clustree'</code>. Then install <code>'scrnaboxR'</code>: </p>
<pre><code>devtools::install_github(&quot;neurobioinfo/scrnabox/scrnaboxR&quot;)
</code></pre>
<p>The <code>'scrnaboxR'</code> is an R package that provides a collection of functions for conducting enrichment analysis and other analyses associated with single-cell RNA sequencing (scRNA-seq) data. It serves as a companion to scrnabox, offering a range of tools and functionalities to enhance scRNA-seq data analysis. You need to add the R info in <code>scrnabox_config.ini</code>, you can define the path of R library in <code>R_LIB_PATH=</code>, version of R in <code>R_VERSION</code>, you can add the path of <code>cell ranger</code>in <code>MODULECELLRANGER</code> </p>
<h2 id="processed-data"><a href="https://github.com/neurobioinfo/scrnabox/blob/main/README_PROC.md">Processed Data</a></h2>
<p>The pipeline can also be used with <a href="https://github.com/neurobioinfo/scrnabox/blob/main/README_PROC.md">processed data</a> from different projects.</p>
<h2 id="tutorial">Tutorial</h2>
<p>To analyze DGE look at <a href="https://github.com/neurobioinfo/scrnabox/blob/main/tutorial/practice.md">Practice</a>. </p>
<h2 id="references">References</h2>
<ul>
<li>Stuart, T., Butler, A., Hoffman, P., Hafemeister, C., Papalexi, E., Mauck III, W. M., ... &amp; Satija, R. (2019). Comprehensive integration of single-cell data. Cell, 177(7), 1888-1902.</li>
</ul>
<h2 id="contributing">Contributing</h2>
<p>This is an early version, any contribute or suggestion is appreciated, you can directly contact with <a href="https://github.com/saeidamiri1">Saeid Amiri</a> or <a href="https://github.com/RhalenaThomas">Rhalena Thomas</a>. </p>
<h2 id="changelog">Changelog</h2>
<p>Every release is documented on the <a href="https://github.com/neurobioinfo/scrnabox/releases">GitHub Releases page</a>.</p>
<h2 id="license">License</h2>
<p>This project is licensed under the MIT License - see the <a href="https://github.com/neurobioinfo/scrnabox/blob/main/LICENSE">LICENSE.md</a> file for details.</p>
<h2 id="acknowledgement">Acknowledgement</h2>
<p>The pipeline is done as part Dark Genome project, it is written by <a href="https://github.com/saeidamiri1">Saeid Amiri</a> with associate of Rhalena Thomas and  Neuro Bioinformatics Core. Copyright belong MNI BIOINFO CORE (https://github.com/neurobioinfo). </p>
<h2 id="todo">Todo</h2>
<p><strong><a href="#contents">â¬† back to top</a></strong></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="README_SCRNA/" class="btn btn-neutral float-right" title="Standard scRNA">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
      <span><a href="README_SCRNA/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme_extra.js" defer></script>
    <script src="js/theme.js" defer></script>
      <script src="search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>

<!--
MkDocs version : 1.4.3
Build Date UTC : 2023-08-17 03:47:17.948612+00:00
-->
