<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/installation/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Installation - scRNAbox</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Installation";
        var mkdocs_page_input_path = "installation.md";
        var mkdocs_page_url = "/installation/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scRNAbox
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Installation</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#scrnaboxslurm-installation">scrnabox.slurm installation</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cellranger-installation">CellRanger installation</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#r-library-preparation-and-r-package-installation">R library preparation and R package installation</a>
    </li>
    </ul>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Overview</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../SCRNA/">- Standard Analysis Track</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../HTO/">- Cell Hashtag Analysis Track</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">- Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">- Outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset1/">- Standard Analysis: Midbrain dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset2/">- Cell Hashtag Analysis: PBMC dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PROC/">- Processed Data</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../library_prep/">- Library preparation</a>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../FAQ/">FAQ</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">- License</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../changelog/">- Changelog</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">- Contributing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">- Acknowledgement</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scRNAbox</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
      <li class="breadcrumb-item active">Installation</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="installation">Installation</h1>
<p>To use the scRNAbox pipeline, the folowing must be installed on your High-Performance Computing (HPC) system:</p>
<ul>
<li><a href="#scrnaboxslurm-installation">scrnabox.slurm</a></li>
<li><a href="#cellranger-installation">CellRanger</a></li>
<li><a href="#r-library-preparation-and-r-package-installation">R and R packages</a></li>
</ul>
<hr />
<h3 id="scrnaboxslurm-installation">scrnabox.slurm installation</h3>
<p><code>scrnabox.slurm</code> is written in bash and can be used with any Slurm system. To download the latest version of <code>scrnabox.slurm</code> (v0.1.35) run the following command: </p>
<pre><code>wget https://github.com/neurobioinfo/scrnabox/releases/download/v0.1.35/scrnabox.slurm.zip
unzip scrnabox.slurm.zip
</code></pre>
<p>For a description of the options for running <code>scrnabox.slurm</code> run the following command:</p>
<pre><code>bash ~/scrnabox.slurm/launch_scrnabox.sh -h 
</code></pre>
<p>If the <code>scrnabox.slurm</code> has been installed properly, the above command should return the folllowing:</p>
<pre><code>        mandatory arguments:
                -d  (--dir)  = Working directory (where all the outputs will be printed) (give full path)
                --steps  =  Specify what steps, e.g., 2 to run just step 2, 2-4, run steps 2 through 4)

        optional arguments:
                -h  (--help)  = See helps regarding the pipeline options. 
                --method  = Choose what scRNA method you want to use; use HTO  and SCRNA for for hashtag nad Standard scRNA, respectively. 
                --nFeature_RNA_L  = Lower threshold of number of unique RNA transcripts for each cell, it filters nFeature_RNA &gt; nFeature_RNA_L.  
                --nFeature_RNA_U  = Upper threshold of number of unique RNA transcripts for each cell, it filters --nFeature_RNA_U.  
                --nCount_RNA_L  = Lower threshold for nCount_RNA, it filters nCount_RNA &gt; nCount_RNA_L   
                --nCount_RNA_U  = Upper threshold for  nCount_RNA, it filters nCount_RNA &lt; nCount_RNA_U  
                --mitochondria_percent_L  = Lower threshold for the amount of mitochondrial transcript, it is in percent, mitochondria_percent &gt; mitochondria_percent_L. 
                --mitochondria_percent_U  = Upper threshold for the amount of mitochondrial transcript, it is in percent, mitochondria_percent &lt; mitochondria_percent_U. 
                --log10GenesPerUMI_U  = Upper threshold for the log number of genes per UMI for each cell, it is in percent,log10GenesPerUMI=log10(nFeature_RNA)/log10(nCount_RNA). mitochondria_percent &lt; log10GenesPerUMI_U. 
                --log10GenesPerUMI_L  = Lower threshold for the log number of genes per UMI for each cell, log10GenesPerUMI=log10(nFeature_RNA)/log10(nCount_RNA). mitochondria_percent &gt; log10GenesPerUMI_L.  
                --msd  = you can get the hashtag labels by running the following code 
                --marker  = Find marker. 
                --sinfo  = Do you need sample info? 
                --fta  = FindTransferAnchors 
                --enrich  = Annotation 
                --dgelist  = creates a DGEListobject from a table of counts obtained from seurate objects. 
                --genotype  = Run the genotype contrast. 
                --celltype  = Run the Genotype-cell contrast. 
                --cont  = You can directly call the contrast to the pipeline.  
                --seulist                = You can directly call the list of seurat objects to the pipeline. 
</code></pre>
<hr />
<h3 id="cellranger-installation">CellRanger installation</h3>
<p>For information regarding the installation of <code>CellRanger</code>, please visit the 10X Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation">documentation</a>. If CellRanger is already installed on your HPC system, you may skip the CellRanger installation procedures.</p>
<hr />
<h3 id="r-library-preparation-and-r-package-installation">R library preparation and R package installation</h3>
<p>Users must first install <code>R</code> onto their HPC system: </p>
<pre><code># install R
module load r/4.2.1
</code></pre>
<p>Then, users should create a designated directory on their HPC system where the required R packages will be installed:</p>
<pre><code># make common R library
mkdir R_library
cd R_library

# open R
R 

# set common R library path
R_LIB_PATH=&quot;~/R_library&quot;
.libPaths(R_LIB_PATH)

# load packages
library(Seurat)
library(ggplot2)
library(dplyr)
library(foreach)
library(doParallel)
library(Matrix)
library(DoubletFinder)
library(cowplot)
library(clustree)
library(xlsx)
library(enrichR)
library(stringi)
library(limma)
library(tidyverse)
library(edgeR)
library(vctrs)
library(RColorBrewer)
library(fossil)
library(openxlsx)
library(stringr)
library(ggpubr)
devtools::install_github(“neurobioinfo/scrnabox/scrnaboxR”)
</code></pre>
<hr />
<p>Upon completing the installation procedures, users can proceed with the scRNAbox pipeline using either the <a href="../SCRNA/">Standard scRNAseq Analysis Track</a> or <a href="../HTO/">Cell Hashtag scRNAseq Analysis Track</a>. </p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href=".." class="btn btn-neutral float-left" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../SCRNA/" class="btn btn-neutral float-right" title="- Standard Analysis Track">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href=".." style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../SCRNA/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
