<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/Step4HTO/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Step 4: Demultiplexing (HTO) - scRNAbox</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Step 4: Demultiplexing (HTO)";
        var mkdocs_page_input_path = "Step4HTO.md";
        var mkdocs_page_url = "/Step4HTO/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scRNAbox
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipeline</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 0: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 1: FASTQ to expression matrix</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step2/">Step 2: Seurat object and ambient RNA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step3/">Step 3: Quality control and filtering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4/">Step 4: Doublet removal (standard)</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Step 4: Demultiplexing (HTO)</a>
    <ul class="current">
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step5/">Step 5: Integration</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step6/">Step 6: Clustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step7/">Step 7: Cluster annotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step8/">Step 8: Differential gene expression</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../config/">Job configurations</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading the Midbrain dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset1/">Standard analysis track: Midbrain dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DEG/">Analysis of DGE outputs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../pbmc_download/">Downloading the PBMC dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset2/">HTO analysis track: PBMC dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../library_prep/">Manual CellRanger library preparation</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">License</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">Contributing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">Acknowledgement</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scRNAbox</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Pipeline &raquo;</li>
      <li class="breadcrumb-item active">Step 4: Demultiplexing (HTO)</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="step-4-demultiplexing-and-doublet-detection-hto-track">Step 4: Demultiplexing and doublet detection (HTO track)</h1>
<p>In Step 4 of the HTO track, Seuratâ€™s implementation (<em>MULTIseqDemux</em>) of the tag assignment algorithm outlined in Multi-seq is used to demultiplex pooled samples and identify doublets according to the expression matrices of the sample-specific barcodes (<a href="https://pubmed.ncbi.nlm.nih.gov/31209384/">McGinnis et al 2019</a>).</p>
<hr />
<p>The following parameters are adjustable for Step 4 (<code>~/working_directory/job_info/parameters/step4_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Default</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to export an RNA expression matrix</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to export a metadata dataframe</td>
</tr>
<tr>
<td style="text-align: left;">par_seurat_object</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">If users already have a Seurat object(s), they may provide the path to a directory that contains an existing Seurat object(s) to initiate the pipeline at Step 4</td>
</tr>
<tr>
<td style="text-align: left;">par_normalization.method</td>
<td style="text-align: left;">CLR</td>
<td style="text-align: left;">Method for normalizing the HTO assay</td>
</tr>
<tr>
<td style="text-align: left;">par_scale.factor</td>
<td style="text-align: left;">1000</td>
<td style="text-align: left;">Scale factor for scaling the HTO assay</td>
</tr>
<tr>
<td style="text-align: left;">par_selection.method</td>
<td style="text-align: left;">vst</td>
<td style="text-align: left;">Method for selecting the most variable features in the HTO assay</td>
</tr>
<tr>
<td style="text-align: left;">par_nfeatures</td>
<td style="text-align: left;">5</td>
<td style="text-align: left;">Number of features to select as top variable features for the HTO assay. This value is dependent on the number of sample specific barcodes used in the experiment</td>
</tr>
<tr>
<td style="text-align: left;">par_dims_umap</td>
<td style="text-align: left;">5</td>
<td style="text-align: left;">Number of dimensions to use as input features for uniform manifold approximation and projection (UMAP) of HTO assay</td>
</tr>
<tr>
<td style="text-align: left;">par_n.neighbor</td>
<td style="text-align: left;">65</td>
<td style="text-align: left;">Number of neighboring points to use in local approximations of manifold structure</td>
</tr>
<tr>
<td style="text-align: left;">par_dimensionality_reduction</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to perform linear dimensionality reduction on the HTO assay</td>
</tr>
<tr>
<td style="text-align: left;">par_npcs_pca</td>
<td style="text-align: left;">30</td>
<td style="text-align: left;">Total Number of principal components to compute and store for principal component analysis (PCA) of HTO assay</td>
</tr>
<tr>
<td style="text-align: left;">par_dropDN</td>
<td style="text-align: left;">Yes</td>
<td style="text-align: left;">Whether or not to remove predicted doublets and negatives from downstream analyses</td>
</tr>
<tr>
<td style="text-align: left;">par_label_dropDN</td>
<td style="text-align: left;">Doublet, Negative</td>
<td style="text-align: left;">Labels used to identify doublet and negative droplets</td>
</tr>
<tr>
<td style="text-align: left;">par_quantile</td>
<td style="text-align: left;">0.9</td>
<td style="text-align: left;">The quantile to use for droplet classification using <em>MULTIseqDemux</em></td>
</tr>
<tr>
<td style="text-align: left;">par_autoThresh</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: left;">Whether or not to perform automated threshold finding to define the best quantile for droplet classification using <em>MULTIseqDemux</em></td>
</tr>
<tr>
<td style="text-align: left;">par_maxiter</td>
<td style="text-align: left;">5</td>
<td style="text-align: left;">Maximum number of iterations to use if autoThresh = TRUE</td>
</tr>
<tr>
<td style="text-align: left;">par_RidgePlot_ncol</td>
<td style="text-align: left;">3</td>
<td style="text-align: left;">Number of columns used to display RidgePlots, which visualizes the enrichment of barcode labels across samples</td>
</tr>
<tr>
<td style="text-align: left;">par_old_antibody_label</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">If you wish to rename the barcode labels, first list the existing barcode labels in this parameter. old antibody labels can be identified in the "_old_antibody_label_MULTIseqDemuxHTOcounts" file produced by running Step 4 msd</td>
</tr>
<tr>
<td style="text-align: left;">par_new_antibody_label</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">If you wish to rename the barcode labels, list the new labels corresponding to the old labels listed in the parameter above</td>
</tr>
</tbody>
</table>
<hr />
<p>To demultiplex the samples and identify doublets, the first step is to obtain the barcode labels used in the analysis by running the following command:</p>
<pre><code>bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 4 \
--msd T 
</code></pre>
<p><strong>Note:</strong> This step will produce the <code>old_antibody_label_MULTIseqDemuxHTOcounts.csv</code> file, which contains the names of the old HTO labels. The names of the HTO labels can be revised to be more descriptive in the execution parameters of this step (par_old_antibody_label; par_new_antibody_label)</p>
<p>Next, demultiplex the samples and identify doublets by running the following command:</p>
<pre><code>bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 4 
</code></pre>
<p>The resulting output files are deposited into <code>~/working_directory/step4</code>. For a description of the outputs see <a href="../outputs/">here</a>.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../Step4/" class="btn btn-neutral float-left" title="Step 4: Doublet removal (standard)"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../Step5/" class="btn btn-neutral float-right" title="Step 5: Integration">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../Step4/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../Step5/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
