<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/Dataset1/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>scRNAbox analysis of midbrain dataset - scRNAbox</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "scRNAbox analysis of midbrain dataset";
        var mkdocs_page_input_path = "Dataset1.md";
        var mkdocs_page_url = "/Dataset1/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scRNAbox
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipeline</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 0: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 1: FASTQ to expression matrix</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step2/">Step 2: Seurat object and ambient RNA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step3/">Step 3: Quality control and filtering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4/">Step 4: Doublet removal (standard)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4HTO/">Step 4: Demultiplexing (HTO)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step5/">Step 5: Integration</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step6/">Step 6: Clustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step7/">Step 7: Cluster annotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step8/">Step 8: Differential gene expression</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Additional information</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../config/">Job configurations</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Outputs</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">scRNAbox analysis of midbrain dataset</a>
    <ul class="current">
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading the midbrain dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../library_prep/">Manual CellRanger library preparation</a>
                  </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../FAQ/">FAQ</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">- License</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../changelog/">- Changelog</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">- Contributing</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">- Acknowledgement</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scRNAbox</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Tutorial &raquo;</li>
      <li class="breadcrumb-item active">scRNAbox analysis of midbrain dataset</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="scrnabox-analysis-of-the-midbrain-dataset">ScRNAbox analysis of the midbrain dataset</h2>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#downloading-the-midbrain-dataset">Downloading the midbrain dataset</a></li>
<li><a href="#installation">Installation</a><ul>
<li><a href="#scrnaboxslurm-installation">scrnabox.slurm installation</a></li>
<li><a href="#cellranger-installation">CellRanger installation</a></li>
<li><a href="#r-library-preparation-and-r-package-installation">R library preparation and R package installation</a></li>
</ul>
</li>
<li><a href="#scrnabox-standard-analysis-track">scRNAbox: Standard Analysis Track</a><ul>
<li><a href="#step-0-pipeline-initiation">Step 0: Pipeline initiation</a></li>
<li><a href="#step-1-fastq-to-gene-expression-matrix">Step 1: FASTQ to gene expression matrix</a>  </li>
<li><a href="#step-2-create-seurat-object-and-remove-ambient-rna">Step 2: Create Seurat object and remove ambient RNA </a>  </li>
<li><a href="#step-3-quality-control-and-filtering">Step 3: Quality control and filtering</a></li>
<li><a href="#step-4-doublet-detection">Step 4: Doublet detection</a> </li>
<li><a href="#step-5-integration-and-linear-dimensional-reduction">Step 5: Integration and linear dimensional reduction</a></li>
<li><a href="#step-6-clustering">Step 6: Clustering</a> </li>
<li><a href="#step-7-cluster-annotation">Step 7: Cluster annotation</a> <ul>
<li><a href="#method-1-cluster-marker-gsea">Method 1: Cluster marker GSEA</a> </li>
<li><a href="#method-2-module-score">Method 2: Module score</a> </li>
<li><a href="#method-3-reference-based-annotation">Method 3: Reference-based annotation</a> </li>
<li><a href="#visualizin-the-expression-of-select-features">Visualizing the expression of select features</a>    </li>
</ul>
</li>
<li><a href="#step-8-differential-gene-expression-contrasts">Step 8: Differential gene expression contrasts</a><ul>
<li><a href="#create-dgelist-object">Create DGEList object</a> </li>
<li><a href="#sample-sample-contrasts">Sample-sample contrasts</a> </li>
<li><a href="#sample-cell-contrasts">Sample-cell contrasts</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#job-configurations">Job Configurations</a>      </li>
</ul>
<hr />
<h2 id="introduction">Introduction</h2>
<p>This guide illustrates the steps taken to analyze the midbrain dataset (<a href="https://academic.oup.com/brain/article/145/3/964/6469020">Smajic et al. 2022</a>) that was presented in our <a href="">pre-print manuscript</a>. This dataset describes single-nuclei transcriptomes from the post-mortem midbrains of five individuals with Parkinsonâ€™s disease (PD) and six controls sequenced separately. </p>
<hr />
<h2 id="downlaoding-the-midbrain-dataset">Downlaoding the midbrain dataset</h2>
<p>In you want to use the midbrain dataset to test the scRNAbox pipeline, please see <a href="../midbrain_download/">here</a> for detialed instructions on how to download the publicly available data.</p>
<hr />
<h2 id="installation">Installation</h2>
<h4 id="scrnaboxslurm-installation">scrnabox.slurm installation</h4>
<p>To download the latest version of <code>scrnabox.slurm</code> (v0.1.35) run the following command: </p>
<pre><code>wget https://github.com/neurobioinfo/scrnabox/releases/download/v0.1.35/scrnabox.slurm.zip
unzip scrnabox.slurm.zip
</code></pre>
<p>For a description of the options for running <code>scrnabox.slurm</code> run the following command:</p>
<pre><code>bash /pathway/to/scrnabox.slurm/launch_scrnabox.sh -h 
</code></pre>
<p>If the <code>scrnabox.slurm</code> has been installed properly, the above command should return the folllowing:</p>
<pre><code>        mandatory arguments:
                -d  (--dir)  = Working directory (where all the outputs will be printed) (give full path)
                --steps  =  Specify what steps, e.g., 2 to run just step 2, 2-4, run steps 2 through 4)

        optional arguments:
                -h  (--help)  = See helps regarding the pipeline arguments. 
                --method  = Select your preferred method: HTO and SCRNA for hashtag, and Standard scRNA, respectively. 
                --msd  = You can get the hashtag labels by running the following code 
                --markergsea  = Identify marker genes for each cluster and run marker gene set enrichment analysis (GSEA) using EnrichR libraries. 
                --knownmarkers  = Run module score and visualize the expression of known cell type marker genes. 
                --referenceannotation  = Run module score and visualize the expression of known cell type marker genes. 
                --annotate  = Run module score and visualize the expression of known cell type marker genes. 
                --addmeta  = Add metadata columns to the Seurat object 
                --rundge  = Perform differential gene expression contrasts 
                --seulist  = You can directly call the list of seurat objects to the pipeline.  

 ------------------- 
 For a comprehensive help, visit https://github.com/neurobioinfo/scrnabox for documentation.
</code></pre>
<hr />
<h4 id="cellranger-installation">CellRanger installation</h4>
<p>For information regarding the installation of CellRanger, please visit the 10X Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation">documentation</a>. If CellRanger is already installed on your HPC system, you may skip the CellRanger installation procedures.</p>
<p>For our analysis of the midbrain dataset we used the 10XGenomics GRCh38-3.0.0 reference genome and CellRanger v5.0.1. For more information regarding how to prepare reference genomes for the CellRanger <em>counts</em> pipeline, please see the 10X Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/release-notes/build#grch38_3.0.0">documentation</a>.</p>
<hr />
<h4 id="r-library-preparation-and-r-package-installation">R library preparation and R package installation</h4>
<p>We must prepapre a common R library where we will load all of the required R packages. If the required R packages are already installed on your HPC system in a common R library, you may skip the following procedures. 
<br /></p>
<p>We will first install <code>R</code>. The analyses presented in our <a href="">pre-print manuscript</a> were conducted using v4.2.1.</p>
<pre><code># install R
module load r/4.2.1
</code></pre>
<p>Then, we will run the installation code, which creates a directory where the R packages will be loaded and will install the required R packages:</p>
<pre><code># Folder for R packages 
R_PATH=~/path/to/R/library
mkdir -p $R_PATH

# Install package
Rscript ./scrnabox.slurm/soft/R/install_packages_scrnabox.R $R_PATH
</code></pre>
<hr />
<h2 id="scrnabox-pipeline">scRNAbox pipeline</h2>
<h3 id="step-0-pipeline-initiation">Step 0: Pipeline initiation</h3>
<p>Now that <code>scrnabox.slurm</code>, <code>CellRanger</code>, <code>R</code>, and the Required R packages have been installed, we can proceed to our analysis with the scRNAbox pipeline. We will create a <code>pipeline</code> folder designated for the analysis and run Step 0, selecting the standard analysis track (<code>--method SCRNA</code>), using the following code:</p>
<pre><code>mkdir pipeline
cd pipeline

export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 0 \
--method SCRNA
</code></pre>
<p>Next, we will navigate to the <code>scrnabox_config.ini</code> file in <code>~/pipeline/job_info/configs</code> to define the path to the R library (<code>R_LIB_PATH=</code>), the version of R (<code>R_VERSION=</code>), and the path to CellRanger (<code>MODULECELLRANGER=</code>):</p>
<pre><code>cd ~/pipeline/job_info/configs
nano scrnabox_config.ini

MODULECELLRANGER=mugqic/cellranger/5.0.1
R_VERSION=4.2.1
R_LIB_PATH=~/R
</code></pre>
<hr />
<h3 id="step-1-fastq-to-gene-expression-matrix">Step 1: FASTQ to gene expression matrix</h3>
<p>In Step 1, we will run the CellRanger <em>counts</em> pipeline to generate feature-barcode expression matrices from the FASTQ files. While it is possible to manually prepare the <code>library.csv</code> files for each of the 11 samples in the experiment prior to running Step 1, for this analysis we are going to opt for automated library preparation. For more information regarding the manual prepartion of <code>library.csv</code> files, please see the the <a href="../library_prep/">CellRanger library preparation</a> tutorial. <br /> 
<br /> 
For our analysis of the midbrain dataset we set the following execution parameters for Step 1 (<code>~/pipeline/job_info/parameters/step1_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_automated_library_prep</td>
<td style="text-align: left;">yes</td>
</tr>
<tr>
<td style="text-align: left;">par_fastq_directory</td>
<td style="text-align: left;">/path/to/directory/with/fastqs</td>
</tr>
<tr>
<td style="text-align: left;">par_sample_names</td>
<td style="text-align: left;">PD1, PD2, PD3, PD4, PD5, CTRL1, CTRL2, CTRL3, CTRL4, CTRL5, CTRL6</td>
</tr>
<tr>
<td style="text-align: left;">par_rename_samples</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_new_sample_names</td>
<td style="text-align: left;">Parkinson1, Parkinson2, Parkinson3, Parkinson4, Parkinson5, Control1, Control2, Control3, Control4, Control5, Control6</td>
</tr>
<tr>
<td style="text-align: left;">par_paired_end_seq</td>
<td style="text-align: left;">TRUE</td>
</tr>
<tr>
<td style="text-align: left;">par_ref_dir_grch</td>
<td style="text-align: left;">~/genome/10xGenomics/refdata-cellranger-GRCh38-3.0.0</td>
</tr>
<tr>
<td style="text-align: left;">par_r1_length</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_include_introns</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_mempercode</td>
<td style="text-align: left;">30</td>
</tr>
</tbody>
</table>
<p><strong>Note:</strong> The parameters file for each step is located in <code>~/pipeline/job_info/parameters</code>. For a comprehensive description of the execution parameters for each see see <a href="../reference/">here</a>. </p>
<p>Given that CellRanger runs a user interface and is not submitted as a Job, it is recommended to run Step 1 in a <strong>'screen'</strong> which will allow the the task to keep running if the connection is broken. To run Step 1, use the following command:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

screen -S run_smajic_application_case
bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 1
</code></pre>
<p>The outputs of the CellRanger <em>counts</em> pipeline are deposited into <code>~/pipeline/step1</code>.</p>
<hr />
<h3 id="step-2-create-seurat-object-and-remove-ambient-rna">Step 2: Create Seurat object and remove ambient RNA</h3>
<p>In Step 2, we are going to use the CellRanger-generated feature-barcode matrices to produce unique Seurat objects for each of the 11 samples. Ambient RNA detection and removal is optional for this step; however, because <a href="https://academic.oup.com/brain/article/145/3/964/6469020">Smajic et al.</a> did not perform this analytical procedure we will skip it. 
<br /> </p>
<p>For our analysis of the midbrain dataset we set the following execution parameters for Step 2 (<code>~/pipeline/job_info/parameters/step2_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_ambient_RNA</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_count_matrices</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_min.cells_L</td>
<td style="text-align: left;">3</td>
</tr>
<tr>
<td style="text-align: left;">par_min.features_L</td>
<td style="text-align: left;">1000</td>
</tr>
</tbody>
</table>
<p>We can run Step 2 using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 2
</code></pre>
<p>Step 2 produces the following outputs for each sample. As an example we show the outputs for Parkinson1.</p>
<pre><code>step2
â”œâ”€â”€ figs2
â”‚   â””â”€â”€ vioplot_Parkinson1.png
â”œâ”€â”€ info2
â”‚   â”œâ”€â”€ Parkinson11_RNA.txt
â”‚   â”œâ”€â”€ MetaDataParkinson11.txt
â”‚   â”œâ”€â”€ MetaDataParkinson1.txt
â”‚   â”œâ”€â”€ meta_infoParkinson1.txt
â”‚   â”œâ”€â”€ Parkinson1_RNA.txt
â”‚   â”œâ”€â”€ sessionInfo.txt
â”‚   â””â”€â”€ summary_Parkinson1.txt
â””â”€â”€ objs2
    â””â”€â”€ Parkinson1.rds
</code></pre>
<p><strong>Note:</strong> For a comprehensive description of the outputs for each Analytical Step, please see the <a href="../outputs/">Outputs</a> section of the scRNAbox documentation.</p>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/2127d7ea-f7be-43dd-9ae0-2421376e2727" width="350" height="100"> <br /> 
</p>

<p><strong>Figure 1. Figure produced by Step 2 of the Standard Analysis Track.</strong> The figure for the Parkinson1 sample is shown as an example. Sample-specific violin plots are produced to visualize the distribution of genes per cell (nFeature_RNA), molecules per cell (nCount_RNA), percentage of mitochondrial-encoded genes per cell (percent.mt), and perentage of ribosomal genes per cell (percent.ribo).</p>
<hr />
<h3 id="step-3-quality-control-and-filtering">Step 3: Quality control and filtering</h3>
<p>In this Step, we are going to perform quality control (QC) procedures and filter out low quality cells. We are going to filter out cells with &lt;1500 unique molecules, &gt;10% mitochondrial-encoded genes, and &gt;10% ribosomal genes. In addition, we are going to remove mitochondrial-encoded and ribosomal genes and will perform cell cycle scoring. Prior to performing cell cycle scoring, we must normalize and scale the counts matrix. </p>
<p>For our analysis of the midbrain dataset we set the following execution parameters for Step 3 (<code>~/pipeline/job_info/parameters/step2_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_seurat_object</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_nFeature_RNA_L</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_nFeature_RNA_U</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_nCount_RNA_L</td>
<td style="text-align: left;">1500</td>
</tr>
<tr>
<td style="text-align: left;">par_nCount_RNA_U</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_mitochondria_percent_L</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_mitochondria_percent_U</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">par_ribosomal_percent_L</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_ribosomal_percent_U</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">par_remove_mitochondrial_genes</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_remove_ribosomal_genes</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_remove_genes</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_normalization.method</td>
<td style="text-align: left;">LogNormalize</td>
</tr>
<tr>
<td style="text-align: left;">par_scale.factor</td>
<td style="text-align: left;">10000</td>
</tr>
<tr>
<td style="text-align: left;">par_selection.method</td>
<td style="text-align: left;">vst</td>
</tr>
<tr>
<td style="text-align: left;">par_nfeatures</td>
<td style="text-align: left;">2500</td>
</tr>
<tr>
<td style="text-align: left;">par_top</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">par_npcs_pca</td>
<td style="text-align: left;">30</td>
</tr>
<tr>
<td style="text-align: left;">par_cells</td>
<td style="text-align: left;">500</td>
</tr>
<tr>
<td style="text-align: left;">par_dims</td>
<td style="text-align: left;">12</td>
</tr>
<tr>
<td style="text-align: left;">par_dims_umap</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">par_n.neighbors</td>
<td style="text-align: left;">65</td>
</tr>
</tbody>
</table>
<p>We can run Step 3 using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 3
</code></pre>
<p>Step 3 produces the following outputs for each sample. As an example we show the outputs for Parkinson1.</p>
<pre><code>step3
â”œâ”€â”€ figs3
â”‚   â”œâ”€â”€ cellcycle_Parkinson1.png
â”‚   â”œâ”€â”€ dimplot_pcaParkinson1.png
â”‚   â”œâ”€â”€ dimplot_umapParkinson1.png
â”‚   â”œâ”€â”€ elbowplotParkinson1.png
â”‚   â”œâ”€â”€ QC_vioplot_Parkinson1.png
â”‚   â””â”€â”€ VariableFeaturePlotParkinson1.png
â”œâ”€â”€ info3
â”‚   â”œâ”€â”€ MetaDataParkinson1.txt
â”‚   â”œâ”€â”€ meta_info_Parkinson1.txt
â”‚   â”œâ”€â”€ most_variable_genes_Parkinson1.txt
â”‚   â”œâ”€â”€ Parkinson1_RNA.txt
â”‚   â”œâ”€â”€ sessionInfo.txt
â”‚   â””â”€â”€ summary_Parkinson1.txt
â””â”€â”€ objs3
    â””â”€â”€ Parkinson1.rds
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/dbbb39e3-68ee-4e40-aed2-3835476386cd" width="800" height="120"> <br /> 
</p>

<p><strong>Figure 2. Figures produced by Step 3 of the Standard Analysis Track.</strong> The figures for the Parkinson1 sample are shown as an example. <strong>A)</strong> Distribution of QC metrics after filtering according to the user-defined thresholds. <strong>B)</strong> Variable features plot showing the top 2500 most variable features; the top 10 most variable features are labelled. <strong>C)</strong> Elbow plot to visualize the percentage of variance explained by each principal component (PC). <strong>D)</strong> Principal component analysis (PCA) visualizing the first two PCs. <strong>E)</strong> Uniform Manifold Approximation and Projections (UMAP) plot, taking the first ten PCs as input. <strong>F)</strong> Distibution of G2M and S scores across cells.</p>
<hr />
<h3 id="step-4-doublet-detection">Step 4: Doublet detection</h3>
<p>In this Step, we are going to identify doublets (erroneous libraries produced by two or more cells) and remove them from downstream analyses using the <a href="https://www.cell.com/cell-systems/pdfExtended/S2405-4712(19)30073-0">DoubletFinder</a> tool (McGinnis et al. 2019). For optimal performance, DoubletFinder requires the user to define the following parameters:</p>
<ul>
<li>The number of statistically significant PCs (par_PCs)</li>
<li>The number of artificial doublets to generate (par_pN)</li>
<li>The expected doublet rate for each sample (par_expected_doublet_rate)</li>
</ul>
<p>The <strong>number of statistically significant PCs</strong> can be informed by the elbow plots produced in Step 2; it this case the top 15 PCs should maintain a robust compression of the data across samples. DoubletFinder is largely invariant to the <strong>number of artifical doublets generated</strong>, therefore we will maintain the default parameter of 0.25. The <strong>expected doublet rate</strong> can be informed by the number of recovered cells (~8% for ~10,000 cells recovered). The number of recovered cells can be informed by the <code>barcodes.tsv.gz</code> file produced by the CellRanger <em>counts</em> pipeline, which is located in <code>~/pipeline/step1/&lt;sample&gt;/output_folder/outs/filtered_feature_bc_matrix</code>. The number of recovered cells for each sample and the corresponding doublet rate is shown below.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Sample</th>
<th style="text-align: left;"># of recovered cells</th>
<th style="text-align: left;">Expected doublet rate (%)</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Control1</td>
<td style="text-align: left;">4863</td>
<td style="text-align: left;">3.9%</td>
</tr>
<tr>
<td style="text-align: left;">Control2</td>
<td style="text-align: left;">4827</td>
<td style="text-align: left;">3.9%</td>
</tr>
<tr>
<td style="text-align: left;">Control3</td>
<td style="text-align: left;">2632</td>
<td style="text-align: left;">2.3%</td>
</tr>
<tr>
<td style="text-align: left;">Control4</td>
<td style="text-align: left;">5221</td>
<td style="text-align: left;">3.9%</td>
</tr>
<tr>
<td style="text-align: left;">Control5</td>
<td style="text-align: left;">3703</td>
<td style="text-align: left;">3.1%</td>
</tr>
<tr>
<td style="text-align: left;">Control6</td>
<td style="text-align: left;">6533</td>
<td style="text-align: left;">5.4%</td>
</tr>
<tr>
<td style="text-align: left;">Parkinson1</td>
<td style="text-align: left;">2512</td>
<td style="text-align: left;">2.3%</td>
</tr>
<tr>
<td style="text-align: left;">Parkinson2</td>
<td style="text-align: left;">6437</td>
<td style="text-align: left;">4.6%</td>
</tr>
<tr>
<td style="text-align: left;">Parkinson3</td>
<td style="text-align: left;">3963</td>
<td style="text-align: left;">3.1%</td>
</tr>
<tr>
<td style="text-align: left;">Parkinson4</td>
<td style="text-align: left;">2495</td>
<td style="text-align: left;">1.6%</td>
</tr>
<tr>
<td style="text-align: left;">Parkinson5</td>
<td style="text-align: left;">5937</td>
<td style="text-align: left;">4.6%</td>
</tr>
</tbody>
</table>
<p>The expected doublet rates are approximations obtained from the 10X Genomics Next GEM Single Cell 3' v3.1 <a href="https://kb.10xgenomics.com/hc/en-us/articles/360059124751-Why-is-the-multiplet-rate-different-for-the-Next-GEM-Single-Cell-3-LT-v3-1-assay-compared-to-other-single-cell-applications-">documentation</a>, which was used by <a href="https://academic.oup.com/brain/article/145/3/964/6469020">Smajic et al.</a> for library preparation. </p>
<p>For our analysis of the midbrain dataset we set the following execution parameters for Step 4 (<code>~/pipeline/job_info/parameters/step4_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_dropDN</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_PCs</td>
<td style="text-align: left;">15</td>
</tr>
<tr>
<td style="text-align: left;">par_pN</td>
<td style="text-align: left;">0.25</td>
</tr>
<tr>
<td style="text-align: left;">par_sct</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr>
<td style="text-align: left;">par_sample_names</td>
<td style="text-align: left;">Control1, Control2, Control3, Control4, Control5, Control6, Parkinson1, Parkinson2, Parkinson3, Parkinson4, Parkinson5</td>
</tr>
<tr>
<td style="text-align: left;">par_expected_doublet_rate</td>
<td style="text-align: left;">0.039, 0.039, 0.023, 0.039, 0.031, 0.054, 0.023, 0.046, 0.031, 0.016, 0.046</td>
</tr>
</tbody>
</table>
<p>We can run Step 4 using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 4
</code></pre>
<p>Step 4 produces the following outputs for each sample. As an example we show the outputs for Parkinson1.</p>
<pre><code>step4
â”œâ”€â”€ figs4
â”‚   â””â”€â”€ Parkinson1DF.classifications.png
â”œâ”€â”€ info4
â”‚   â”œâ”€â”€ meta_info_Parkinson1.txt
â”‚   â”œâ”€â”€ Parkinson1_RNA.txt
â”‚   â”œâ”€â”€ sessionInfo.txt
â”‚   â””â”€â”€ seu_MetaDataParkinson1.txt
â””â”€â”€ objs4
    â””â”€â”€ Parkinson1.rds
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/d736992c-51dc-4c8c-af8b-d9ad0d5ef860" width="350" height="100"> <br /> 
</p>

<p><strong>Figure 3. Figure produced by Step 4 of the Standard Analysis Track.</strong> The figure for the Parkinson1 sample is shown as an example.  Uniform Manifold Approximation and Projections (UMAP) plot showing the cell-type classification (singlet or doublet) for each droplet. In the figure title, the first value represents the number of simulated droplets (0.25), the second value represents the neighbourhood size (0.03), and the third value represents the number of predicted doublets (19). </p>
<hr />
<h3 id="step-5-integration-and-linear-dimensional-reduction">Step 5: Integration and linear dimensional reduction</h3>
<p>In this Step, we are going to integrate the individual Seurat objects to enable joint analyses across all 11 samples. We will then perform normalization, scaling and linear dimensional reduction on the integrated assay. The outputs from Step 5 will inform the optimal clustering parameters for Step 6. </p>
<p>For our analysis of the midbrain dataset we set the following execution parameters for Step 5:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_skip_integration</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_FindIntegrationAnchors_dim</td>
<td style="text-align: left;">25</td>
</tr>
<tr>
<td style="text-align: left;">par_DefaultAssay</td>
<td style="text-align: left;">RNA</td>
</tr>
<tr>
<td style="text-align: left;">par_normalization.method</td>
<td style="text-align: left;">LogNormalize</td>
</tr>
<tr>
<td style="text-align: left;">par_selection.method</td>
<td style="text-align: left;">vst</td>
</tr>
<tr>
<td style="text-align: left;">par_nfeatures</td>
<td style="text-align: left;">4000</td>
</tr>
<tr>
<td style="text-align: left;">par_RunUMAP_n.neighbors</td>
<td style="text-align: left;">65</td>
</tr>
<tr>
<td style="text-align: left;">par_RunPCA_npcs</td>
<td style="text-align: left;">30</td>
</tr>
<tr>
<td style="text-align: left;">par_RunUMAP_dims</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_jackstraw</td>
<td style="text-align: left;">Yes</td>
</tr>
</tbody>
</table>
<p>We can run Step 5 using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 5
</code></pre>
<p>Step 5 produces the following outputs.</p>
<pre><code>step5
â”œâ”€â”€ figs5
â”‚   â”œâ”€â”€ DimPlot_pca.png
â”‚   â”œâ”€â”€ DimPlot_umap.png
â”‚   â”œâ”€â”€ elbow.png
â”‚   â””â”€â”€ Jackstraw_plot.png
â”œâ”€â”€ info5
â”‚   â”œâ”€â”€ meta_info_seu_step5.csv
â”‚   â”œâ”€â”€ sessionInfo.txt
â”‚   â”œâ”€â”€ seu_int_MetaData.txt
â”‚   â””â”€â”€ seu_int_RNA.txt
â””â”€â”€ objs5
    â””â”€â”€ seu_step5.rds
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/3142e6c8-67e7-4b59-afec-99fcc43ee15f" width="550" height="100"> <br /> 
</p>

<p><strong>Figure 4. Figures produced by Step 5 of the Standard Analysis Track.</strong> <strong>A)</strong> Principal component analysis (PCA) visualizing the first two principal components (PC) of the integrated assay. <strong>B)</strong> Uniform Manifold Approximation and Projections (UMAP) plot of the integrated assay, taking the first ten PCs as input. <strong>C)</strong> Jackstraw plot to visualize the distribution of p-values for each PC. <strong>D)</strong> Elbow plot to visualize the percentage of variance explained by each PC.</p>
<hr />
<h3 id="step-6-clustering">Step 6: Clustering</h3>
<p>In this Step, we will cluster the cells to indentify groups of cells with similar expression profiles. Based on the Elbow and Jackstraw plots produced in Step 5, we are going to use the first 25 PCs for the nearest-neighbour graph construction and to run the Uniform Manifold Approximation and Projection (UMAP) dimensional reduction. We will cluster the cells at a clustering resolution of 0.1 to 1.5, in intervals of 1.5. To determine the stability of clusters, we will run the Louvain clustering algorithm five times for each clustering resolution, while shuffling the order of the nodes in the graph for each iteration, and will compute the Adjusted Rand Index (ARI) between pairs of clusters at a given clustering resolution.</p>
<p>For our analysis of the midbrain dataset we set the following execution parameters for Step 6:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_skip_step5</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_FindNeighbors_dims</td>
<td style="text-align: left;">25</td>
</tr>
<tr>
<td style="text-align: left;">par_RunUMAP_dims</td>
<td style="text-align: left;">25</td>
</tr>
<tr>
<td style="text-align: left;">par_FindNeighbors_k.param</td>
<td style="text-align: left;">30</td>
</tr>
<tr>
<td style="text-align: left;">par_FindNeighbors_prune.SNN</td>
<td style="text-align: left;">1/15</td>
</tr>
<tr>
<td style="text-align: left;">par_FindClusters_resolution</td>
<td style="text-align: left;">0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0,1.1,1.2,1.3,1.4,1.5</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_ARI</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_RI_reps</td>
<td style="text-align: left;">5</td>
</tr>
</tbody>
</table>
<p>We can run Step 6 using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 6
</code></pre>
<p>Step 6 produces the following outputs.</p>
<pre><code>step6
â”œâ”€â”€ ARI
â”‚   â”œâ”€â”€ ARI.png
â”‚   â””â”€â”€ test.xlsx
â”œâ”€â”€ figs6
â”‚   â”œâ”€â”€ clustree_int.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.1.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.2.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.3.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.4.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.5.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.6.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.7.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.8.png
â”‚   â”œâ”€â”€ integrated_snn_res.0.9.png
â”‚   â”œâ”€â”€ integrated_snn_res.1.1.png
â”‚   â”œâ”€â”€ integrated_snn_res.1.2.png
â”‚   â”œâ”€â”€ integrated_snn_res.1.3.png
â”‚   â”œâ”€â”€ integrated_snn_res.1.4.png
â”‚   â”œâ”€â”€ integrated_snn_res.1.5.png
â”‚   â””â”€â”€ integrated_snn_res.1.png
â”œâ”€â”€ info6
â”‚   â”œâ”€â”€ meta_info.csv
â”‚   â”œâ”€â”€ sessionInfo.txt
â”‚   â”œâ”€â”€ seu_MetaData.txt
â”‚   â””â”€â”€ seu_RNA.txt
â””â”€â”€ objs6
    â””â”€â”€ seu_step6.rds
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/7a0e2e2f-acb6-4a0d-9e4a-579651ce4d8b" width="750" height="100"> <br /> 
</p>

<p><strong>Figure 5. Figures produced by Step 6 of the Standard Analysis Track.</strong> <strong>A)</strong> ClustTree plot to visualize inter-cluster dynamics at varying cluster resolutions. <strong>B)</strong> Mean (top panel) and standard deviation (sd; middle panel) of the Adjusted RNA Index (ARI) between clustering pairs at each user-defined clustering resolution. The bottom panel shows the number of clusters at each user-defined clustering resolution. <strong>C)</strong> Uniform Manifold Approximation and Projections (UMAP) plot at a clustering resolution of 1.5. </p>
<hr />
<h3 id="step-7-cluster-annotation">Step 7: Cluster annotation</h3>
<p>In this Step, we are going to annotate the clusters identified in Step 6 to define the cellular species in the midbrain dataset. scRNAbox provides three distinct methods for cluster annotations</p>
<ul>
<li><strong>Method 1:</strong> Cluster marker gene set enrichment analysis (GSEA)</li>
<li><strong>Method 2:</strong> Module score</li>
<li><strong>Method 2:</strong> Reference-based annotation</li>
</ul>
<p>In addition to these three Methods, we can <strong>visualize the expression of select features</strong> to further inform the cellular species in the dataset.</p>
<p>For comprehensive description of each cluster annotation Method, please see the <a href="../SCRNA/">Standard scRNAseq Analysis Track</a> section of the scRNAbox documentation or our pre-print manuscript. </p>
<hr />
<h4 id="method-1-cluster-marker-gsea">Method 1: Cluster marker GSEA</h4>
<p>Using Method 1, we are first going to identify differentially expressed marker genes for each cluster. We must define the number of marker genes for each cluster that we want scRNAbox to report and selecte a clustering resolution that we want to annotate. In this case we will report the top five marker genes for each cluster at a clustering resolution of 1.5.</p>
<p>To identify the marker genes for each cluster, we set the following execution parameters for Step 7 (<code>step7_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_level_cluster</td>
<td style="text-align: left;">integrated_snn_res.1.9</td>
</tr>
<tr>
<td style="text-align: left;">par_top_sel</td>
<td style="text-align: left;">5</td>
</tr>
<tr>
<td style="text-align: left;">par_db</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_module_score</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_module_score</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_reference</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_level_celltype</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_FindTransferAnchors_dim</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_futureglobalsmaxSize</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_visualize_select_features</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_select_features</td>
<td style="text-align: left;">NULL</td>
</tr>
</tbody>
</table>
<p>We can identify the marker genes for each cluster using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 7 \
--marker T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step7
â”œâ”€â”€ figs7
â”‚   â”œâ”€â”€ marker
â”‚   â”‚   â””â”€â”€ heatmap.pdf
â”‚   â”œâ”€â”€ module_score
â”‚   â”œâ”€â”€ umap.pdf
â”‚   â”œâ”€â”€ umap_splitted.pdf
â”‚   â””â”€â”€ visualize_select_features
â”œâ”€â”€ info7
â”‚   â”œâ”€â”€ marker
â”‚   â”‚   â”œâ”€â”€ cluster_just_genes.xlsx
â”‚   â”‚   â”œâ”€â”€ ClusterMarkers.csv
â”‚   â”‚   â”œâ”€â”€ ClusterMarkers.rds
â”‚   â”‚   â”œâ”€â”€ cluster_whole.xlsx
â”‚   â”‚   â””â”€â”€ top_sel.csv
â”‚   â”œâ”€â”€ module_score
â”‚   â””â”€â”€ sessionInfo_marker.txt
â””â”€â”€ objs7
</code></pre>
<p>In addition to identifying the marker genes for each cluster, the above code produces UMAP plots at the user-defined clustering resolution (1.5) to visualize the clustering landscape across all cells in the dataset and at the sample level. </p>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/3e338a94-7248-4d10-b904-367b065482c4" width="750" height="100"> <br /> 
</p>

<p><strong>Figure 6. Uniform Manifold Approximation and Projections (UMAP) plots at the user-defined clustering resolution.</strong> <strong>A)</strong> The clustering landscape at the user-defined clustering resolution across all cells in the dataset. <strong>B</strong>) The clustering landscape at the user-defined clustering resolution, stratified by sample. </p>
<p>Now that we have identified the marker genes for each cluster, we will perform a <strong>gene set enrichment analysis (GSEA)</strong>; we will test the differentially expressed genes (DEG) in the positive direction (Log2 fold-change &gt; 0.00) for enrichment across gene set libraries that define cell types using the EnrichR tool. For this analysis, we will leverage the following libraries:</p>
<ul>
<li>Descartes_Cell_Types_and_Tissue_2021; </li>
<li>CellMarker_Augmented_2021; </li>
<li>Azimuth_Cell_Types_2021 cell type libraries.</li>
</ul>
<p>To perform GSEA, we set the following execution parameters for Step 7 (<code>step7_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_level_cluster</td>
<td style="text-align: left;">integrated_snn_res.1.5</td>
</tr>
<tr>
<td style="text-align: left;">par_top_sel</td>
<td style="text-align: left;">5</td>
</tr>
<tr>
<td style="text-align: left;">par_db</td>
<td style="text-align: left;">Descartes_Cell_Types_and_Tissue_2021,<br /> CellMarker_Augmented_2021,<br />Azimuth_Cell_Types_2021</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_module_score</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_module_score</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_reference</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_level_celltype</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_FindTransferAnchors_dim</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_futureglobalsmaxSize</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_visualize_select_features</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_select_features</td>
<td style="text-align: left;">NULL</td>
</tr>
</tbody>
</table>
<p>If your HPC <strong>allows access to the internet</strong>, we can perform GSEA using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 7 \
--enrich T
</code></pre>
<p><strong>Note:</strong> If your HPC <strong>does not allow access to the internet</strong>, you will have to run GSEA locally. For more information, please see the Standard scRNAseq <a href="../SCRNA/">documentation</a> under the Step 7 section.</p>
<p>The above code produces the following outputs. As an example, we are only showing the outputs for cluster 0.</p>
<pre><code>step7
â””â”€â”€ annot_enrich
    â”œâ”€â”€ cluster0
    â”‚   â”œâ”€â”€ Er.genes.1.csv
    â”‚   â”œâ”€â”€ Er.genes.2.csv
    â”‚   â”œâ”€â”€ Er.genes.3.csv
    â”‚   â”œâ”€â”€ plotenrich1.pdf
    â”‚   â”œâ”€â”€ plotenrich2.pdf
    â”‚   â””â”€â”€ plotenrich3.pdf
    â”œâ”€â”€ cluster1
    â”œâ”€â”€ cluster2
    â”œâ”€â”€ cluster3
    â”œâ”€â”€ cluster4
    â”œâ”€â”€ cluster5
    â”œâ”€â”€ cluster6
    â”œâ”€â”€ cluster7
    â”œâ”€â”€ cluster8
    â”œâ”€â”€ cluster9
    â”œâ”€â”€ cluster10
    â”œâ”€â”€ cluster11
    â”œâ”€â”€ cluster12
    â”œâ”€â”€ cluster13
    â”œâ”€â”€ cluster14
    â”œâ”€â”€ cluster15
    â”œâ”€â”€ cluster16
    â”œâ”€â”€ cluster17
    â”œâ”€â”€ cluster18
    â”œâ”€â”€ cluster19
    â”œâ”€â”€ cluster20
    â”œâ”€â”€ cluster21
    â””â”€â”€ cluster22
</code></pre>
<p>After performing cluster marker GSEA and curating the results, we can produce our first iteration of the cluster annotations.<br /></p>
<p><strong>Note:</strong> Visualizing intermediate cluster annotations is not incorporated into the scRNAbox pipeline; however we provide the code to do so below. Once users are satisfied with their final cluster annotations, they can provide the curated results in the parameters file for <a href="">Step 8</a>, which will be discussed below. </p>
<pre><code>## load and open R
module load r/4.2.1
R

## load parameters
# path to common R library
r_lib_path = &quot;~/R&quot;
# path to pipeline directory
output_dir = &quot;~/pipeline&quot;
# clustering resolution to cluster
clustering_resolution = &quot;integrated_snn_res.1.5&quot;
# intermediate cluster annotations
intermediate_cluster_labels = c(&quot;Oligodendrocytes&quot;, &quot;Oligodendrocytes&quot;, &quot;Neuron&quot;,&quot;Oligodendrocytes&quot;,&quot;Oligodendrocytes&quot;,&quot;Oligodendrocytes&quot;,&quot;Oligodendrocytes&quot;,&quot;Oligodendrocytes&quot;,&quot;OPC&quot;, &quot;Endothelial cells&quot;,&quot;Microglia&quot;, &quot;Oligodendrocytes&quot;,&quot;Astrocytes&quot;, &quot;Neuron&quot;, &quot;Astrocytes&quot;, &quot;Endothelial cells&quot;,&quot;Endothelial cells&quot;,  &quot;Astrocytes&quot;, &quot;Neuron&quot;,&quot;Neuron&quot;,&quot;Neuron&quot;, &quot;Microglia&quot;, &quot;Astrocytes&quot;)


## load library
.libPaths(r_lib_path)
packages&lt;-c('Seurat','ggplot2', 'dplyr','stringi','limma','tidyverse','edgeR')
lapply(packages, library, character.only = TRUE)

## load Step 6 Seurat RDS object
sample_name&lt;-list.files(path = paste(output_dir, &quot;/step6/objs6&quot;,sep=&quot;&quot;),pattern = &quot;*.rds&quot;)
seu.int.c&lt;-readRDS(paste(output_dir,'/step6/objs6/',sample_name, sep=''))

## set cluster annotations obtained from cluster annotations
cluster.ids&lt;-intermediate_cluster_labels

## set cluster resolution and rename cluster identities
seu.int.c &lt;- SetIdent(seu.int.c, value = clustering_resolution)
names(cluster.ids) &lt;- levels(seu.int.c)    
seu.int.c &lt;- RenameIdents(seu.int.c, cluster.ids) 

##plot UMAP
DimPlot(seu.int.c, reduction = &quot;umap&quot;, label = TRUE, pt.size = 0.5) + NoLegend()
ggsave(file = paste(output_dir,'/step7/figs7','/intermediate_cluster_annotation.pdf', sep=''))
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/1c0d4342-3c8b-454d-bed8-fe71f31128b0" width="500" height="100"> <br /> 
</p>

<p><strong>Figure 7. Figures produced by Method 1 (Cluster Marker GSEA) of the scRNAbox cluster annotation module.</strong> <strong>A)</strong> The top expressional markers that define each cluster are visualized through a heatmap showing the expression  across cells, stratified by cluster. <strong>B)</strong> Differentially expressed marker genes in the positive direction (Log2 fold-change &gt; 0.00) can be tested for enrichment across gene-set libraries that define cell types using the EnrichR tool. The enrichment results are visualized through a bar plot which displays the 20 most enriched terms for a particular cluster. As an example, we show the enrichment results of cluster 0 using the Azimuth_Cell_Types_2021 cell type library. <strong>C)</strong> Uniform Manifold Approximation and Projections (UMAP) plot showing the intermediate cluster annotations. </p>
<hr />
<h4 id="visualizing-the-expression-of-select-features">Visualizing the expression of select features</h4>
<p>Now that we have broadly defined the cellular species that comprise our clusters, we are going to explore the expression of the marker genes used by <a href="https://academic.oup.com/brain/article/145/3/964/6469020">Smajic et al.</a> to define their clusters:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Cell type</th>
<th style="text-align: left;">Gene</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Oligodendrocytes</td>
<td style="text-align: left;"><em>MOBP</em></td>
</tr>
<tr>
<td style="text-align: left;">OPC</td>
<td style="text-align: left;"><em>VCAN</em></td>
</tr>
<tr>
<td style="text-align: left;">Astrocytes</td>
<td style="text-align: left;"><em>AQP4</em></td>
</tr>
<tr>
<td style="text-align: left;">Ependymal</td>
<td style="text-align: left;"><em>FOXJ1</em></td>
</tr>
<tr>
<td style="text-align: left;">Microglia</td>
<td style="text-align: left;"><em>CD74</em></td>
</tr>
<tr>
<td style="text-align: left;">Endothelial</td>
<td style="text-align: left;"><em>CLDN5</em></td>
</tr>
<tr>
<td style="text-align: left;">Pericytes</td>
<td style="text-align: left;"><em>PDGFRB</em></td>
</tr>
<tr>
<td style="text-align: left;">Excitatory neurons</td>
<td style="text-align: left;"><em>SLC17A6</em></td>
</tr>
<tr>
<td style="text-align: left;">Inhibitory neurons</td>
<td style="text-align: left;"><em>GAD2</em></td>
</tr>
<tr>
<td style="text-align: left;">GABAergic neurons</td>
<td style="text-align: left;"><em>GAD2</em>, <em>GRIK1</em></td>
</tr>
<tr>
<td style="text-align: left;">Dopaminergic neurons (DaN)</td>
<td style="text-align: left;"><em>TH</em></td>
</tr>
<tr>
<td style="text-align: left;">Degenerating DaN</td>
<td style="text-align: left;"><em>CADPS2</em></td>
</tr>
</tbody>
</table>
<p>To visualize these features, we set the following execution parameters for Step 7 (<code>step7_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_level_cluster</td>
<td style="text-align: left;">integrated_snn_res.1.5</td>
</tr>
<tr>
<td style="text-align: left;">par_top_sel</td>
<td style="text-align: left;">5</td>
</tr>
<tr>
<td style="text-align: left;">par_db</td>
<td style="text-align: left;">Descartes_Cell_Types_and_Tissue_2021,<br /> CellMarker_Augmented_2021,<br />Azimuth_Cell_Types_2021</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_module_score</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_module_score</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_reference</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_level_celltype</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_FindTransferAnchors_dim</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_futureglobalsmaxSize</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_visualize_select_features</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_select_features</td>
<td style="text-align: left;">MOBP, VCAN, AQP4,FOXJ1, CD74, CLDN5, PDGFRB, SLC17A6, GAD2, GRIK1, TH, CADPS2</td>
</tr>
</tbody>
</table>
<p>We can visualize the expression of these features using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 7 \
--marker T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step7
â””â”€â”€ figs7
    â””â”€â”€ visualize_select_features
        â”œâ”€â”€select_feature_dot_plot.pdf 
        â”œâ”€â”€select_feature_feature_plot.pdf
        â””â”€â”€select_feature_violin_plot.pdf
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/f6150d60-fcf4-4426-9e56-7e60f94eedd1" width="900" height="100"> <br /> 
</p>

<p><strong>Figure 8. Figures produced by the "visualize select features" option in the scRNAbox cluster annotation module.</strong>  The expression of select features can be visualized at the cluster level via <strong>A)</strong> a dot plot and <strong>B)</strong> violin plots. <strong>C)</strong> The expression of select features can be visualized at the cell level via feature plots.</p>
<p>Based on the results of the above analyses, we can re-visit our cluster annotations using the same intermediate annotation code as above and visualize the annotations via a UMAP.</p>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/d682e588-0e51-4c8b-832e-384dc428a467" width="400" height="100"> <br /> 
</p>

<p><strong>Figure 9. Intermediate cluster annotations.</strong> Uniform Manifold Approximation and Projections (UMAP) plot showing the intermediate cluster annotations after leveraging the "visualize select feature" function of scRNAbox's cluster annotation module.</p>
<hr />
<h4 id="method-2-module-score">Method 2: Module score</h4>
<p>Using Method 2, we are going to comparatively quantify the expression of gene sets across clusters at the single-cell level. We will first define the gene set that we want to explore in a csv file; as an example, we are going to explore the expression of some well-known marker genes for the cellular species of interest. </p>
<p>We will first produce a csv file with the following structure. This csv can be found <strong>HERE</strong></p>
<table>
<thead>
<tr>
<th style="text-align: left;">da_neurons</th>
<th style="text-align: left;">NPC_orStemLike</th>
<th style="text-align: left;">mature_neurons</th>
<th style="text-align: left;">excitatory_neurons</th>
<th style="text-align: left;">inhbitory_neurons</th>
<th style="text-align: left;">astrocytes</th>
<th style="text-align: left;">oligodendrocytes</th>
<th style="text-align: left;">radial_glia</th>
<th style="text-align: left;">epithelial</th>
<th style="text-align: left;">microglia</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">TH</td>
<td style="text-align: left;">DCX</td>
<td style="text-align: left;">RBFOX3</td>
<td style="text-align: left;">GRIA2</td>
<td style="text-align: left;">GAD1</td>
<td style="text-align: left;">GFAP</td>
<td style="text-align: left;">MBP</td>
<td style="text-align: left;">PTPRC</td>
<td style="text-align: left;">HES1</td>
<td style="text-align: left;">IBA1</td>
</tr>
<tr>
<td style="text-align: left;">SLC6A3</td>
<td style="text-align: left;">NEUROD1</td>
<td style="text-align: left;">SYP</td>
<td style="text-align: left;">GRIA1</td>
<td style="text-align: left;">GAD2</td>
<td style="text-align: left;">S100B</td>
<td style="text-align: left;">MOG</td>
<td style="text-align: left;">AIF1</td>
<td style="text-align: left;">HES5</td>
<td style="text-align: left;">P2RY12</td>
</tr>
<tr>
<td style="text-align: left;">SLC18A2</td>
<td style="text-align: left;">TBR1</td>
<td style="text-align: left;">VAMP1</td>
<td style="text-align: left;">GRIA4</td>
<td style="text-align: left;">GAT1</td>
<td style="text-align: left;">AQP4</td>
<td style="text-align: left;">OLIG1</td>
<td style="text-align: left;">ADGRE1</td>
<td style="text-align: left;">SOX2</td>
<td style="text-align: left;">P2RY13</td>
</tr>
<tr>
<td style="text-align: left;">SOX6</td>
<td style="text-align: left;">PCNA</td>
<td style="text-align: left;">VAMP2</td>
<td style="text-align: left;">GRIN1</td>
<td style="text-align: left;">PVALB</td>
<td style="text-align: left;">APOE</td>
<td style="text-align: left;">OLIG2</td>
<td style="text-align: left;">VIM</td>
<td style="text-align: left;">SOX10</td>
<td style="text-align: left;">TREM119</td>
</tr>
<tr>
<td style="text-align: left;">NDNF</td>
<td style="text-align: left;">MKI67</td>
<td style="text-align: left;">TUBB3</td>
<td style="text-align: left;">GRIN2B</td>
<td style="text-align: left;">GABR2</td>
<td style="text-align: left;">SOX9</td>
<td style="text-align: left;">SOX10</td>
<td style="text-align: left;">TNC</td>
<td style="text-align: left;">NES</td>
<td style="text-align: left;">GPR34</td>
</tr>
<tr>
<td style="text-align: left;">SNCG</td>
<td style="text-align: left;">SOX2</td>
<td style="text-align: left;">SYT1</td>
<td style="text-align: left;">GRIN2A</td>
<td style="text-align: left;">GABR1</td>
<td style="text-align: left;">SLC1A3</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">PTPRZ1</td>
<td style="text-align: left;">CDH1</td>
<td style="text-align: left;">SIGLECH</td>
</tr>
<tr>
<td style="text-align: left;">ALDH1A1</td>
<td style="text-align: left;">NES</td>
<td style="text-align: left;">BSN</td>
<td style="text-align: left;">GRIN3A</td>
<td style="text-align: left;">GBRR1</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">FAM107A</td>
<td style="text-align: left;">NOTCH1</td>
<td style="text-align: left;">TREM2</td>
</tr>
<tr>
<td style="text-align: left;">CALB1</td>
<td style="text-align: left;">PAX6</td>
<td style="text-align: left;">HOMER1</td>
<td style="text-align: left;">GRIN3</td>
<td style="text-align: left;">GABRB2</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">HOPX</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">CX3CR1</td>
</tr>
<tr>
<td style="text-align: left;">TACR2</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">SLC17A6</td>
<td style="text-align: left;">GRIP1</td>
<td style="text-align: left;">GABRB1</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">LIFR</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">FCRLS</td>
</tr>
<tr>
<td style="text-align: left;">SLC17A6</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">CAMK2A</td>
<td style="text-align: left;">GABRB3</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">ITGB5</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">OLFML3</td>
</tr>
<tr>
<td style="text-align: left;">SLC32A1</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">GABRA6</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">IL6ST</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">HEXB</td>
</tr>
<tr>
<td style="text-align: left;">OTX2</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">GABRA1</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">SLC1A3</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">TGFBR1</td>
</tr>
<tr>
<td style="text-align: left;">GRP</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">GABRA4</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">SALL1</td>
</tr>
<tr>
<td style="text-align: left;">LPL</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">TRAK2</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">MERTK</td>
</tr>
<tr>
<td style="text-align: left;">CCK</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">PROS1</td>
</tr>
<tr>
<td style="text-align: left;">VIP</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
<p>We can then define the location of our csv file in the execution parameters for Step 7 (<code>step7_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_level_cluster</td>
<td style="text-align: left;">integrated_snn_res.1.5</td>
</tr>
<tr>
<td style="text-align: left;">par_top_sel</td>
<td style="text-align: left;">5</td>
</tr>
<tr>
<td style="text-align: left;">par_db</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_module_score</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_module_score</td>
<td style="text-align: left;">~/pipeline/module_score.csv</td>
</tr>
<tr>
<td style="text-align: left;">par_reference</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_level_celltype</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_FindTransferAnchors_dim</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_futureglobalsmaxSize</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_visualize_select_features</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_select_features</td>
<td style="text-align: left;">NULL</td>
</tr>
</tbody>
</table>
<p>We can compute the module score for our gene sets using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 7 \
--marker T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step7
â”œâ”€â”€ figs7
â”‚   â””â”€â”€ module_score
â”‚       â”œâ”€â”€module_score_astrocytes.png
â”‚       â”œâ”€â”€module_score_epithelial.png
â”‚       â”œâ”€â”€module_score_da_neurons.png
â”‚       â”œâ”€â”€module_score_excitatory_neurons.png
â”‚       â”œâ”€â”€module_score_inhbitory_neurons.png
â”‚       â”œâ”€â”€module_score_mature_neurons.png
â”‚       â”œâ”€â”€module_score_microglia.png
â”‚       â”œâ”€â”€module_score_NPC_orStemLike.png
â”‚       â”œâ”€â”€module_score_oligodendrocytes.png
â”‚       â””â”€â”€module_score_radial_glia.png
â””â”€â”€ info7
    â””â”€â”€module_score
       â””â”€â”€geneset_by_cluster.csv

</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/d42626d0-9c41-49e8-a3a9-8f38c0f91345" width="900" height="100"> <br /> 
</p>

<p><strong>Figure 10. Figures produced by Method 2 (Module score) of the scRNAbox cluster annotation module.</strong> Uniform Manifold Approximation and Projections (UMAP) plots showing the module score across established marker genes for <strong>A)</strong> dopaminergic neurons, <strong>B)</strong> neural progenitor cells, <strong>C)</strong> mature neurons, <strong>D)</strong> excitatory neurons, <strong>E)</strong> inhibitory neurons, <strong>F)</strong> astrocytes, <strong>G)</strong> oligodendrocytes, <strong>H)</strong> radial glia, <strong>I)</strong> epithelial cells, and <strong>J)</strong> microglia.</p>
<hr />
<h4 id="method-3-reference-based-annotation">Method 3: Reference-based annotation</h4>
<p>Using Method 3, we are going to leverage the cell-type annotations from a reference Seurat object and generate annotation predictions for the query dataset. For reference-based annotation we must define the path to a our reference Seurat object and the column of the reference Seurat object's metadata that contains the cell type annotations. For the midbrain dataset, we are going to use a reference Seurat object from <a href="https://www.nature.com/articles/s41593-022-01061-1">Kamath et al.</a>.</p>
<p>To perform reference-based annotations, we set the following execution parameters for Step 7 (<code>step7_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">Value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_level_cluster</td>
<td style="text-align: left;">integrated_snn_res.1.5</td>
</tr>
<tr>
<td style="text-align: left;">par_top_sel</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_db</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_compute_module_score</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_module_score</td>
<td style="text-align: left;">NULL</td>
</tr>
<tr>
<td style="text-align: left;">par_reference</td>
<td style="text-align: left;">~/reference_seurat_object.rds</td>
</tr>
<tr>
<td style="text-align: left;">par_level_celltype</td>
<td style="text-align: left;">Cell_Type</td>
</tr>
<tr>
<td style="text-align: left;">par_FindTransferAnchors_dim</td>
<td style="text-align: left;">10</td>
</tr>
<tr>
<td style="text-align: left;">par_futureglobalsmaxSize</td>
<td style="text-align: left;">60000 * 1024^2</td>
</tr>
<tr>
<td style="text-align: left;">par_visualize_select_features</td>
<td style="text-align: left;">No</td>
</tr>
<tr>
<td style="text-align: left;">par_select_features</td>
<td style="text-align: left;">NULL</td>
</tr>
</tbody>
</table>
<p>We can perform reference-based annotations using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 7 \
--fta T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step7
â”œâ”€â”€ figs7
â”‚   â””â”€â”€ reference_based_annotation
â”‚       â””â”€â”€UMAP_transferred_labels.pdf
â””â”€â”€ objs7
    â””â”€â”€seu_step7.rds
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/00f64588-79d8-4d77-909f-2251f4381860" width="400" height="100"> <br /> 
</p>

<p><strong>Figure 11. Figure produced by Method 3 (reference-based annotation) of the scRNAbox cluster annotation module.</strong> Uniform Manifold Approximation and Projections (UMAP) plots showing the cluster annotations from the reference Seurat object projected onto the query Seurat object.</p>
<hr />
<h3 id="step-8-differential-gene-expression-contrasts">Step 8: Differential gene expression contrasts</h3>
<p>In this step we are going to perform differential gene expression (DGE) analysis between our samples. ScRNAbox faciltates DGE contrasts between samples (<strong>sample-sample contrasts</strong>) and between samples, stratified by cell type (<strong>sample-cell contrasts</strong>). The DGE contrasts module contains three components:</p>
<p>1) <strong>Create DGEList object</strong> <br />
2) <strong>Sample-sample contrasts</strong><br />
3) <strong>Sample-cell contrasts</strong><br /></p>
<hr />
<h4 id="create-dgelist-object">Create DGEList object</h4>
<p>First, we are going to create a DGElist object. Before doing so, we must define our desired clustering resolution and the final cluster annotations informed by Step 7. We are also going to rename our samples in order to faciliate DGE contrasts.</p>
<p>To create a DGElist object, we set the following execution parameters for Step 8 (<code>step8_par.txt</code>):</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">value</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">par_save_RNA</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_save_metadata</td>
<td style="text-align: left;">Yes</td>
</tr>
<tr>
<td style="text-align: left;">par_level_cluster</td>
<td style="text-align: left;">integrated_snn_res.1.5</td>
</tr>
<tr>
<td style="text-align: left;">par_step8_clus_label</td>
<td style="text-align: left;">Oligodendrocytes, Oligodendrocytes, Excitatory_Neurons, Oligodendrocytes, Oligodendrocytes, Oligodendrocytes, Oligodendrocytes, Oligodendrocytes, OPC, Endothelial_cells, Microglia,  Oligodendrocytes, Astrocytes, Excitatory_Neurons, Astrocytes, Pericytes, Endothelial_cells, Ependymal, GABAergic_neurons, Oligodendrocytes, Inhibitory_neurons, Oligodendrocytes, OPC</td>
</tr>
<tr>
<td style="text-align: left;">par_new_genotype</td>
<td style="text-align: left;">yes</td>
</tr>
<tr>
<td style="text-align: left;">par_old_sample_label</td>
<td style="text-align: left;">Control1, Control2, Control3, Control4, Control5, Control6, Parkinson1, Parkinson2, Parkinson3, Parkinson4, Parkinson5</td>
</tr>
<tr>
<td style="text-align: left;">par_new_sample_label</td>
<td style="text-align: left;">Control, Control, Control, Control, Control, Control, Parkinson, Parkinson, Parkinson, Parkinson, Parkinson</td>
</tr>
</tbody>
</table>
<p><strong>Note:</strong> Cell names and sample names cannot have spaces. For example, do not write "Endothelial cells", instead write "Endothelial_cells".</p>
<p>We can create the DGElist object using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 8 \
--dgelist T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step8
â”œâ”€â”€ figs8
â”‚   â””â”€â”€ final_cluster_annotation.pdf
â”œâ”€â”€ info8
â”‚   â”œâ”€â”€ de_genes.rds
â”‚   â”œâ”€â”€ dge.rds
â”‚   â”œâ”€â”€ meta_info_de_genes.txt
â”‚   â”œâ”€â”€ meta_info_dge.txt
â”‚   â””â”€â”€ meta_info_seu_step8.txt
â””â”€â”€ objs8
    â””â”€â”€ seu_step8.rds
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/13a7dc2f-56b2-4a0d-b034-fb50089582a9" width="400" height="100"> <br /> 
</p>

<p><strong>Figure 12. Final cluster annotations used for differential gene expression (DGE) contrasts .</strong> Uniform Manifold Approximation and Projections (UMAP) plots showing the final cluster annotation obtained by curating the results from scRNAbox's cluster annotation module (Step 7). The final cluster annotations will be used throughout the DGE contrasts module. </p>
<hr />
<h4 id="sample-sample-contrasts">Sample-sample contrasts</h4>
<p>Now that we have our DGElist object, we can perform DGE contrasts between samples (sample-sample contrasts). As an example, we will test for DGE between Parkinson's disease samples and controls. We must first define our contrast matrix in the sample-sample contrasts parameters file (<code>step8_contrast_genotype.txt</code>):</p>
<pre><code>cont_name control versus
design1 Control Parkinson
</code></pre>
<p>We can perform sample-sample DGE contrasts using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 8 \
--genotype T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step8
â””â”€â”€ cont_genotype
    â””â”€â”€ design1.csv 
</code></pre>
<hr />
<h4 id="sample-cell-contrasts">Sample-cell contrasts</h4>
<p>We can also perform DGE contrasts between samples, stratified by cell type (sample-cell contrasts). As an example, we will test for DGE between microglia from Parkinson's disease and controls. We must first define our contrast matrix in the sample-cell contrasts parameters file (<code>step8_contrast_celltype.txt</code>):</p>
<pre><code>cont_name cell control versus
design1_cell Microglia Control Parkinson
</code></pre>
<p>We can perform sample-cel DGE contrasts using the following code:</p>
<pre><code>export SCRNABOX_HOME=~/scrnabox/scrnabox.slurm
export SCRNABOX_PWD=~/pipeline

bash $SCRNABOX_HOME/launch_scrnabox.sh \
-d ${SCRNABOX_PWD} \
--steps 8 \
--celltype T
</code></pre>
<p>The above code produces the following outputs:</p>
<pre><code>step8
â””â”€â”€ cont_celltype
    â””â”€â”€ design1.csv 
</code></pre>
<hr />
<h3 id="step-9-enrichment-analysis">Step 9: Enrichment analysis</h3>
<hr />
<h3 id="job-configurations">Job Configurations</h3>
<p>The following job configurations were used for our analysis of the midbrain dataset. Job Configurations can be modified for each Analytical Step in the <code>scrnabox_config.ini</code> file in <code>~/pipeline/job_info/configs</code></p>
<table>
<thead>
<tr>
<th style="text-align: left;">Step</th>
<th style="text-align: left;">THREADS_ARRAY</th>
<th style="text-align: left;">MEM_ARRAY</th>
<th style="text-align: left;">WALLTIME_ARRAY</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Step2</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">16g</td>
<td style="text-align: left;">00-05:00</td>
</tr>
<tr>
<td style="text-align: left;">Step3</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">16g</td>
<td style="text-align: left;">00-05:00</td>
</tr>
<tr>
<td style="text-align: left;">Step4</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">45g</td>
<td style="text-align: left;">00-05:00</td>
</tr>
<tr>
<td style="text-align: left;">Step5</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">45g</td>
<td style="text-align: left;">00-05:00</td>
</tr>
<tr>
<td style="text-align: left;">Step6</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">16g</td>
<td style="text-align: left;">00-05:00</td>
</tr>
<tr>
<td style="text-align: left;">Step7 marker</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">40g</td>
<td style="text-align: left;">00-01:00</td>
</tr>
<tr>
<td style="text-align: left;">Step7 fta</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">150g</td>
<td style="text-align: left;">00-09:00</td>
</tr>
<tr>
<td style="text-align: left;">Step8 dgelist</td>
<td style="text-align: left;">4</td>
<td style="text-align: left;">40g</td>
<td style="text-align: left;">00-12:00</td>
</tr>
<tr>
<td style="text-align: left;">Step8 cont</td>
<td style="text-align: left;">10</td>
<td style="text-align: left;">40g</td>
<td style="text-align: left;">00-12:00</td>
</tr>
</tbody>
</table>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../outputs/" class="btn btn-neutral float-left" title="Outputs"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../midbrain_download/" class="btn btn-neutral float-right" title="Downloading the midbrain dataset">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../outputs/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../midbrain_download/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
