<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://neurobioinfo.github.io/DEG/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Analysis of DGE outputs - scRNAbox</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Analysis of DGE outputs";
        var mkdocs_page_input_path = "DEG.md";
        var mkdocs_page_url = "/DEG/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scRNAbox
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipeline</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step0/">Step 0: Set up</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step1/">Step 1: FASTQ to expression matrix</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step2/">Step 2: Seurat object and ambient RNA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step3/">Step 3: Quality control and filtering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4/">Step 4: Doublet removal (standard)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step4HTO/">Step 4: Demultiplexing (HTO)</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step5/">Step 5: Integration</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step6/">Step 6: Clustering</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step7/">Step 7: Cluster annotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Step8/">Step 8: Differential gene expression</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../config/">Job configurations</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../reference/">Execution parameters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../outputs/">Outputs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Log/">Intepreting Log Files</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tutorial</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../midbrain_download/">Downloading the Midbrain dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset1/">Standard analysis track: Midbrain dataset</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Analysis of DGE outputs</a>
    <ul class="current">
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../pbmc_download/">Downloading the PBMC dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Dataset2/">HTO analysis track: PBMC dataset</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../library_prep/">Manual CellRanger library preparation</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">About</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../LICENSE/">License</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../contributing/">Help and Feedback</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Acknowledgement/">Acknowledgement</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scRNAbox</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Tutorial &raquo;</li>
      <li class="breadcrumb-item active">Analysis of DGE outputs</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="analysis-of-dge-outputs">Analysis of DGE outputs</h2>
<p>This tutorial provides the code used for producing Figure 5 of our <a href="https://www.biorxiv.org/content/10.1101/2023.11.13.566851v1">pre-print manuscript</a>; downstream analyses of the differential gene expression (DGE) results for the midbrain dataset (<a href="https://academic.oup.com/brain/article/145/3/964/6469020">Smajic et al. 2022</a>). The tutorial can be broken down into three sections: <br /></p>
<ul>
<li><a href="#1-deg-summary">DEG summary</a><br /></li>
<li><a href="#2-cell-based-dge-enrichment-analysis">Cell-based DGE enrichment analysis</a><br /></li>
<li><a href="#3-sample-based-dge-enrichment-analysis">Sample-based DGE enrichment analysis</a>.<br /></li>
</ul>
<p>The outputs from Step 8 of the <a href="../Dataset1/">scRNAbox analuysis of the midbrain dataset</a> tutorial are used as input.</p>
<hr />
<h2 id="set-up">Set up</h2>
<p><strong>Set seed</strong></p>
<pre><code>set.seed(1234)
</code></pre>
<p><strong>Load libraries</strong></p>
<pre><code>library(ggrepel)
library(ggplot2)
library(stringr)
BiocManager::install("clusterProfiler")
BiocManager::install("pathview")
BiocManager::install("enrichplot")
library(clusterProfiler)
library(enrichplot)
organism = "org.Hs.eg.db"
BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
require(DOSE)
library(stringr)
library(dplyr)
</code></pre>
<hr />
<h2 id="1-deg-summary">1. DEG summary</h2>
<h3 id="11-load-cell-based-all-cells-deg-file">1.1. load cell-based all cells DEG file</h3>
<pre><code>## load DEG csv file
all &lt;- read.delim("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/wilcoxon_all_cells/HCvPD/HCvPD_DEG.csv", header = T, sep = ",")
</code></pre>
<h3 id="12-load-cell-based-cell-type-groups-deg-files">1.2. load cell-based cell type groups DEG files</h3>
<pre><code>## list of cell types used for DEG contrasts
cell_based_cell_types &lt;- c('Astrocytes', 'DaN', 'Endothelial', 'Ependymal', 'Excitatory', 'GABA', 'Inhibitory', 'Microglia', 'Oligocendrocytes', 'OPC', 'Pericytes')

## create empty list object
list &lt;- list()

## load DEG csv files
for(i in unique(cell_based_cell_types)){
  list[i] &lt;- list(read.delim(paste0("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/wilcoxon_celltype_groups/", i,"PDvHC/",i,"PDvHC_DEG.csv"), header = T, sep = ","))
}

## create data frame for each cell type
list2env(list,envir=.GlobalEnv)
</code></pre>
<h3 id="13-load-sample-based-all-cells-deg-file">1.3. load sample-based all cells DEG file</h3>
<pre><code>## load DEG csv file
all_sample &lt;- read.delim("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/pseudo_bulk_all_cells/PDvControl/DGE_AllCellsMainContrast HC vs PD.csv", header = T, sep = ",")
</code></pre>
<h3 id="14-load-sample-based-cell-type-groups-deg-files">1.4. load sample-based cell type groups DEG files</h3>
<pre><code>## list of cell types used for DEG contrasts
sample_based_cell_types &lt;- c('astro', 'DaN', 'endo', 'ependymal', 'excit', 'GABA', 'inhib', 'mg', 'Olig', 'opc', 'peri')

## create empty list object
list &lt;- list()

## load DEG csv files
for(i in unique(sample_based_cell_types)){
  list[paste0(i,"_sample")] &lt;- list(read.delim(paste0("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/pseudo_bulk_celltype_groups/PDvControlbulk/info/DGE_",i,"MainContrast HC vs PD.csv"), header = T, sep = ","))
}

## create data frame for each cell type
list2env(list,envir=.GlobalEnv)
</code></pre>
<h3 id="15-clean-up-cell-based-deg-files">1.5. Clean up cell-based DEG files</h3>
<pre><code>## make list of cell type dataframes
CellType_df &lt;- list(Astrocytes, DaN, Endothelial, Ependymal, Excitatory, GABA, Inhibitory, Microglia, Oligocendrocytes, OPC, Pericytes, all)

## set names of list as cell types
names(CellType_df) &lt;- c('Astrocytes', 'DaN', 'Endothelial', 'Ependymal', 'Excitatory', 'GABA', 'Inhibitory', 'Microglia', 'Oligocendrocytes', 'OPC', 'Pericytes', "all")

## list of columns that we want to keep
cols_keep &lt;- c("X","p_val","avg_log2FC","p_val_adj")

## process each dataframe to only keep DEGs with a p-value &lt; 0.05 and log2FC &gt; 1 or &lt; -1
list &lt;- list()

for (i in unique(names(CellType_df))){
  j &lt;- data.frame(CellType_df[i])
  colnames(j) &lt;- c("X", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")
  j &lt;- j[,colnames(j) %in% cols_keep]
  j &lt;- subset(j, p_val &lt;= 0.05)
  j &lt;- subset(j, avg_log2FC &gt;= 1 | avg_log2FC &lt;= -1)
  if (nrow(j &gt; 0)){
  j$cell_type &lt;- paste0(i)
  j$method &lt;- "cell_based"
  list[i] &lt;- list(j)
  }else{
    print(paste0("did not find any DEGs for ", i))
  }
  list[i]
}

## create data frame for each cell type
list2env(list,envir=.GlobalEnv)
</code></pre>
<h3 id="16-clean-up-sample-based-deg-files">1.6. Clean up sample-based DEG files</h3>
<pre><code>## make list of cell type dataframes
CellType_df &lt;- list(astro_sample, DaN_sample, endo_sample, ependymal_sample, excit_sample, GABA_sample, inhib_sample, mg_sample, Olig_sample, opc_sample, peri_sample, all_sample)

## set names of list as cell types
names(CellType_df) &lt;- c('astro_sample', 'DaN_sample', 'endo_sample', 'ependymal_sample', 'excit_sample', 'GABA_sample', 'inhib_sample', 'mg_sample', 'Olig_sample', 'opc_sample', 'peri_sample', "all_sample")

## list of columns that we want to keep
cols_keep &lt;- c("X","pvalue","log2FoldChange","padj")

## process each dataframe to only keep DEGs with a p-value and log2FC &gt; 1 or &lt; -1
list &lt;- list()

for (i in unique(names(CellType_df))){
  j &lt;- data.frame(CellType_df[i])
  colnames(j) &lt;- c('X', 'baseMean', 'log2FoldChange', 'lfcSE' ,'stat', 'pvalue', 'padj')
  j &lt;- j[,colnames(j) %in% cols_keep]
  j &lt;- subset(j, pvalue &lt;= 0.05)
  j &lt;- subset(j, log2FoldChange &gt;= 1 | log2FoldChange &lt;= -1)
  if (nrow(j &gt; 0)){
  j$cell_type &lt;- paste0(i)
  j$method &lt;- "sample_based"
  list[i] &lt;- list(j)
  }else{
    print(paste0("did not find any DEGs for ", i))
  }
  list[i]
}

## create data frame for each cell type
list2env(list,envir=.GlobalEnv)
</code></pre>
<h3 id="17-compute-number-of-degs-identified-for-each-type-using-cell-based-dge-analysis">1.7. Compute number of DEGs identified for each type using cell-based DGE analysis</h3>
<pre><code>## bind dataframes
# do not include any cell types that identified 0 DEGS (e.g. OPC)
bind_cell_based &lt;- rbind(Astrocytes, DaN, Endothelial, Ependymal, Excitatory, GABA, Inhibitory, Microglia, Oligocendrocytes,Pericytes, all)

## define colours
bind_cell_based$col[bind_cell_based$avg_log2FC &gt; 0] &lt;- "indianred3"
bind_cell_based$col[bind_cell_based$avg_log2FC &lt; 0] &lt;- "dodgerblue2"
bind_cell_based$col[bind_cell_based$avg_log2FC &gt; 0 &amp; bind_cell_based$p_val_adj &lt; 0.05] &lt;- "red4"
bind_cell_based$col[bind_cell_based$avg_log2FC &lt; 0 &amp; bind_cell_based$p_val_adj &lt; 0.05] &lt;- "navy"
bind_cell_based$direction[bind_cell_based$avg_log2FC &gt; 0] &lt;- "up"
bind_cell_based$direction[bind_cell_based$avg_log2FC &lt; 0] &lt;- "down"

## plot
cell_based_DEG_counts &lt;- ggplot(bind_cell_based, aes(x = direction, fill = col)) +
  geom_bar() +
  theme_classic() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.ticks.x = element_blank(),
        strip.background = element_rect(colour=NA, fill=NA),
        strip.text = element_text(size = 12),
        legend.position = "none",
        axis.title.y = element_text(face="bold", size = 12),
        axis.title.x = element_text(face="bold", size = 12),
        plot.title = element_text(hjust = 0.5, face = "bold", size = 12)) +
  scale_fill_manual(values = c("dodgerblue2", "indianred3", "navy","red4")) +
  facet_wrap(~cell_type,  ncol=12, strip.position = "bottom") +
  ylab("Number of DEGs") +
  xlab("Cell type") + 
  scale_y_continuous(limits = c(0, 60)) +
  ggtitle ("Cell-based: MAST")
cell_based_DEG_counts
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/0fc6bd1e-89de-400b-8970-ff2e404d43ac"> <br /> 
</p>

<p><strong>Figure 1. Number of DEGs identified by cell-based DGE analysis for each cell type.</strong> DEGs were calculated using cells as replicates and the MAST framework. Bar chart showing the number of DEGs identified with a log 2 fold-change &lt; -1 (blue) and &gt; 1 (red) and p-values &lt; 0.05. Bonferroni adjusted p-values &lt; 0.05 are indicated by the darker shade.</p>
<h3 id="18-compute-number-of-degs-identified-for-each-cell-type-using-sample-based-dge-analysis">1.8. Compute number of DEGs identified for each cell type using sample-based DGE analysis</h3>
<pre><code>## bind dataframes
bind_sample_based &lt;- rbind(astro_sample, DaN_sample, endo_sample, ependymal_sample, excit_sample, GABA_sample, inhib_sample, mg_sample, Olig_sample, opc_sample, peri_sample, all_sample)

## define colours
bind_sample_based$col[bind_sample_based$log2FoldChange &gt; 0] &lt;- "indianred3"
bind_sample_based$col[bind_sample_based$log2FoldChange &lt; 0] &lt;- "dodgerblue2"
bind_sample_based$col[bind_sample_based$log2FoldChange &gt; 0 &amp; bind_sample_based$padj &lt; 0.05] &lt;- "red4"
bind_sample_based$col[bind_sample_based$log2FoldChange &lt; 0 &amp; bind_sample_based$padj &lt; 0.05] &lt;- "navy"
bind_sample_based$direction[bind_sample_based$log2FoldChange &gt; 0] &lt;- "up"
bind_sample_based$direction[bind_sample_based$log2FoldChange &lt; 0] &lt;- "down"

## rename cell types
bind_sample_based$cell_type[bind_sample_based$cell_type == "astro_sample"] &lt;- "Astrocyte"
bind_sample_based$cell_type[bind_sample_based$cell_type == "DaN_sample"] &lt;- "DaN"
bind_sample_based$cell_type[bind_sample_based$cell_type == "endo_sample"] &lt;- "Endothelial"
bind_sample_based$cell_type[bind_sample_based$cell_type == "ependymal_sample"] &lt;- "Ependymal"
bind_sample_based$cell_type[bind_sample_based$cell_type == "excit_sample"] &lt;- "Excitatory"
bind_sample_based$cell_type[bind_sample_based$cell_type == "GABA_sample"] &lt;- "GABA"
bind_sample_based$cell_type[bind_sample_based$cell_type == "inhib_sample"] &lt;- "Inhibitory"
bind_sample_based$cell_type[bind_sample_based$cell_type == "mg_sample"] &lt;- "Microglia"
bind_sample_based$cell_type[bind_sample_based$cell_type == "Olig_sample"] &lt;- "Oligoden."
bind_sample_based$cell_type[bind_sample_based$cell_type == "opc_sample"] &lt;- "OPC"
bind_sample_based$cell_type[bind_sample_based$cell_type == "peri_sample"] &lt;- "Pericyte"
bind_sample_based$cell_type[bind_sample_based$cell_type == "all_sample"] &lt;- "All Cells"


## plot
sample_based_DEG_counts &lt;- ggplot(bind_sample_based, aes(x = direction, fill = col)) +
  geom_bar() +
  theme_classic() + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.ticks.x = element_blank(),
        strip.background = element_rect(colour=NA, fill=NA),
        strip.text = element_text(size = 12),
        legend.position = "none",
        axis.title.y = element_text(face="bold", size = 12),
        axis.title.x = element_text(face="bold", size = 12),
        plot.title = element_text(hjust = 0.5, face = "bold", size = 12)) +
  scale_fill_manual(values = c("dodgerblue2", "indianred3", "navy","red4")) +
  facet_wrap(~cell_type,  ncol=12, strip.position = "bottom") +
  ylab("Number of DEGs") +
  xlab("Cell type") + 
  ggtitle ("Sample-based: DESeq2")
sample_based_DEG_counts
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/86e4f2dc-6cd7-4b96-bbab-5a60193ad744"> <br /> 
</p>

<p><strong>Figure 2. Number of DEGs identified by sample-based DGE analysis for each cell type.</strong>. DEGs were calculated using samples as replicates and the DESeq2 framework. Bar chart showing the number of DEGs identified with a log 2 fold-change &lt; -1 (blue) and &gt; 1 (red) and p-values &lt; 0.05. Bonferroni adjusted p-values &lt; 0.05 are indicated by the darker shade.</p>
<h3 id="19-deg-overlap-analysis">1.9. DEG overlap analysis</h3>
<pre><code>## list of cell-based dataframes
cell_based_dfs &lt;- list(Astrocytes, DaN, Endothelial, Ependymal, Excitatory, GABA, Inhibitory, Microglia, Oligocendrocytes, OPC, Pericytes, all)

## set names of named list
names(cell_based_dfs) &lt;- c('Astrocytes_CellBased', 'DaN_CellBased', 'Endothelial_CellBased', 'Ependymal_CellBased', 'Excitatory_CellBased', 'GABA_CellBased', 'Inhibitory_CellBased', 'Microglia_CellBased', 'Oligocendrocytes_CellBased', 'OPC_CellBased', 'Pericytes_CellBased', 'all_CellBased')

## list of sample-based dataframes
sample_based_dfs &lt;- list(astro_sample, DaN_sample, endo_sample, ependymal_sample, excit_sample, GABA_sample, inhib_sample, mg_sample, Olig_sample, opc_sample, peri_sample, all_sample)

## set names of named list
names(sample_based_dfs) &lt;- c('Astrocytes_SampleBased', 'DaN_SampleBased', 'Endothelial_SampleBased', 'Ependymal_SampleBased', 'Excitatory_SampleBased', 'GABA_SampleBased', 'Inhibitory_SampleBased', 'Microglia_SampleBased', 'Oligocendrocytes_SampleBased', 'OPC_SampleBased', 'Pericytes_SampleBased', 'all_SampleBased')

## combine the list
combined_list &lt;- append(cell_based_dfs,sample_based_dfs )

## list of unique cell types
cell_types &lt;- unique(str_extract(names(combined_list), "[^_]+"))

## compute overlap 
list &lt;- list()

for (i in unique(cell_types)){
  df &lt;- data.frame(names(combined_list))
  df$num &lt;- rownames(df)
  df_lim &lt;- df[grep(i, df$names.combined_list.), ]

  cell_num &lt;- as.numeric(df_lim$num[grep("CellBased", df_lim$names.combined_list.)])
  sample_num &lt;- as.numeric(df_lim$num[grep("SampleBased", df_lim$names.combined_list.)])

  cell &lt;- data.frame(combined_list[as.numeric(cell_num)])
  colnames(cell) &lt;- c('X', 'p_val', 'avg_log2FC', 'p_val_adj', 'cell_type', 'method')
  sample &lt;- data.frame(combined_list[as.numeric(sample_num)])
  colnames(sample) &lt;- c("X", "log2FoldChange", "pvalue", "padj", "cell_type", "method" )

  # both padjusted
  cell_temp &lt;- cell$X[cell$p_val_adj &lt;= 0.05]
  cell_temp &lt;-  cell_temp[!is.na(cell_temp)]
  sample_temp  &lt;- sample$X[sample$padj &lt;= 0.05]
  sample_temp &lt;-  sample_temp[!is.na(sample_temp)]
  length_adj &lt;- length(intersect(sample_temp, cell_temp))

  # both pvalue
  cell_temp &lt;- cell$X[cell$p_val &lt;= 0.05]
  cell_temp &lt;-  cell_temp[!is.na(cell_temp)]
  sample_temp  &lt;- sample$X[sample$pvalue &lt;= 0.05]
  sample_temp &lt;-  sample_temp[!is.na(sample_temp)]
  length_p &lt;- length(intersect(sample_temp, cell_temp))

  ## cell-based pvalue
  cell_temp &lt;- cell$X
  cell_temp &lt;-  cell_temp[!is.na(cell_temp)]
  length_cell_base &lt;- length(cell_temp)

  ## sample-based pvalue
  sample_temp &lt;- sample$X
  sample_temp &lt;-  sample_temp[!is.na(sample_temp)]
  length_sample_base &lt;- length(sample_temp)

  ## cell-based padjust
  cell_temp &lt;- cell$X[cell$p_val_adj &lt;= 0.05]
  cell_temp &lt;-  cell_temp[!is.na(cell_temp)]
  length_cell_adj &lt;- length(cell_temp)

  ## sample-based padjust
  sample_temp &lt;- sample$X[sample$padj &lt;= 0.05]
  sample_temp &lt;-  sample_temp[!is.na(sample_temp)]
  length_sample_adj &lt;- length(sample_temp)

  ## dataframe
  class &lt;- c('AdjustedP_Overlap', 'P_Overlap', 'P_CellBased', 'P_SampleBased', 'AdjustedP_CellBased',    'AdjustedP_SampleBased')
  val &lt;- c(length_adj,length_p,length_cell_base,length_sample_base, length_cell_adj,length_sample_adj)
  temp_df &lt;- data.frame(class, val)
  temp_df$celltype &lt;- i
  list[i] &lt;- list(temp_df)
}

## set list elements to dataframe
list2env(list,envir=.GlobalEnv)

## bind dataframe
df &lt;- bind_rows(mget(cell_types))

## set fator level
df$class &lt;- factor(df$class, levels = c("P_CellBased", "P_SampleBased", "P_Overlap", "AdjustedP_CellBased", "AdjustedP_SampleBased", "AdjustedP_Overlap"))

## plot
ggplot(df, aes(x = celltype, y = class, fill = val)) +geom_tile(col = "black", fill = "white") + theme_bw()+ geom_text(aes(label = val))+
  theme( axis.text = element_text(size = 12),
         axis.title = element_text(size = 12, face = "bold"),
         legend.title = element_text(face = "bold"),
         axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
         axis.text.y = element_text(colour = "black"))+
  scale_colour_manual(values = c("black", "black"), legend) +
  scale_y_discrete(expand = c(0,0)) + #labels = c("Pseudo-bulk: p-value &lt; 0.05",
  scale_x_discrete(expand = c(0,0)) +
  xlab("Cell Type") +
  ylab("")  + guides(colour = "none")
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/ec4bc949-776c-4dff-9d80-99ffa8c08c7c"> <br /> 
</p>

<p><strong>Figure 3. Number of DEG identified by cell-based DGE-MAST, sample-based DGE-DESeq2, or both frameworks across all cell types</strong>. Only DEGs with olg 2 fold-change &lt; -1 and &gt; 1 are included.</p>
<hr />
<h2 id="2-cell-based-dge-enrichment-analysis">2. Cell-based DGE enrichment analysis</h2>
<h3 id="21-load-cell-based-all-cells-deg-file">2.1. load cell-based all cells DEG file</h3>
<pre><code>## load DEG csv file
all &lt;- read.delim("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/wilcoxon_all_cells/HCvPD/HCvPD_DEG.csv", header = T, sep = ",")
</code></pre>
<h3 id="22-load-cell-based-cell-type-groups-deg-files">2.2. load cell-based cell type groups DEG files</h3>
<pre><code>## list of cell types used for DEG contrasts
cell_based_cell_types &lt;- c('Astrocytes', 'DaN', 'Endothelial', 'Ependymal', 'Excitatory', 'GABA', 'Inhibitory', 'Microglia', 'Oligocendrocytes', 'OPC', 'Pericytes')

## create empty list object
list &lt;- list()

## load DEG csv files
for(i in unique(cell_based_cell_types)){
  list[i] &lt;- list(read.delim(paste0("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/wilcoxon_celltype_groups/", i,"PDvHC/",i,"PDvHC_DEG.csv"), header = T, sep = ","))
}

## create data frame for each cell type
list2env(list,envir=.GlobalEnv)
</code></pre>
<h3 id="23-clean-up-cell-based-deg-files">2.3. Clean up cell-based DEG files</h3>
<pre><code>## make list of cell type dataframe
CellType_df &lt;- list(Astrocytes, DaN, Endothelial, Ependymal, Excitatory, GABA, Inhibitory, Microglia, Oligocendrocytes, OPC, Pericytes, all)

## set names of list as cepp types
names(CellType_df) &lt;- c('Astrocytes', 'DaN', 'Endothelial', 'Ependymal', 'Excitatory', 'GABA', 'Inhibitory', 'Microglia', 'Oligocendrocytes', 'OPC', 'Pericytes', "all")

## list of columns that we want to keep
cols_keep &lt;- c("X","p_val","avg_log2FC","p_val_adj")

## process each dataframe to only keep DEGs with a p-value &lt; 0.05 and log2FC &gt; 1
list_CellBased &lt;- list()

for (i in unique(names(CellType_df))){
  j &lt;- data.frame(CellType_df[i])
  colnames(j) &lt;- c("X", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj")
  j &lt;- j[,colnames(j) %in% cols_keep]
  j &lt;- subset(j, p_val &lt;= 0.05)
  j &lt;- subset(j, avg_log2FC &gt;= 1 | avg_log2FC &lt;= -1)
  if (nrow(j &gt; 0)){
  j$cell_type &lt;- paste0(i)
  j$method &lt;- "cell_based"
  list_CellBased[i] &lt;- list(j)
  }else{
    print(paste0("did not find any DEGs for ", i))
  }
  list_CellBased[i]
}

## create data frame for each cell type
list2env(list_CellBased,envir=.GlobalEnv)
</code></pre>
<h3 id="24-enrichr-cell-type-groups">2.4. EnrichR: Cell type groups</h3>
<pre><code>## do not run enrichr for all cell types
test &lt;- data.frame(names(list_CellBased))
test &lt;- test %&gt;% 
  filter(!grepl('all', names.list_CellBased.))
cell_type_num &lt;- as.numeric(rownames(test))

## create empty list
CellBased_gse_list &lt;- list()

## GSE loop
for (i in cell_type_num){
  ## create cell type specific dataframe
  j &lt;- data.frame(list_CellBased[i])
  colnames(j) &lt;- c('X', 'p_val', 'avg_log2FC', 'p_val_adj', 'cell_type', 'method')
  ## we want the log2 fold change 
  gene_list &lt;- j$avg_log2FC
  ## assign gene to log2fc
  names(gene_list) &lt;- j$X
  ## omit any NA values 
  gene_list&lt;-na.omit(gene_list)
  ## sort the list in decreasing order
  gene_list = sort(gene_list, decreasing = TRUE)
  ## perform gseGO
  gse_bulk&lt;- gseGO(geneList=gene_list, 
                            ont ="BP", 
                            keyType = "SYMBOL", 
                            nPerm = 1000, 
                            minGSSize = 3, 
                            maxGSSize = 800, 
                            pvalueCutoff = 1.00, 
                            verbose = TRUE, 
                            OrgDb = organism, 
                            pAdjustMethod = "none")
  ## organize results
  if (nrow(gse_bulk@result) &gt; 0){
  ## count the gene number
  gene_count&lt;- gse_bulk@result %&gt;% group_by(ID) %&gt;% summarise(count = sum(str_count(core_enrichment, "/")) + 1)
  ## merge with the original dataframe
  dot_df&lt;- left_join(gse_bulk@result, gene_count, by = "ID") %&gt;% mutate(GeneRatio = count/setSize)
  dot_df &lt;- dot_df[order(dot_df$pvalue),]
  dot_df$Celltype &lt;- names(list_CellBased[i])
  ## store in list 
  CellBased_gse_list[i] &lt;- list(dot_df)
  }else{
    print(paste0('no term enriched under specific pvalueCutoff for ',names(list_CellBased[i])))
  }
}

## remove NULL (cell types with no enrichment results)
CellBased_gse_list &lt;- CellBased_gse_list[!sapply(CellBased_gse_list,is.null)]

## set names of list element to cell types
names &lt;- list()
for (i in 1:length(CellBased_gse_list)){
j &lt;- data.frame(CellBased_gse_list[i])
names[i] &lt;- unique(j$Celltype)
}
names(CellBased_gse_list) &lt;- names

## set to dataframe
list2env(CellBased_gse_list,envir=.GlobalEnv)
</code></pre>
<h3 id="25-enrichr-all-cells">2.5. EnrichR: All cells</h3>
<pre><code>## retrieve logFC
original_gene_list &lt;- all$avg_log2FC
## assign logFC to gene
names(original_gene_list) &lt;- all$X
## omit any NA values 
gene_list&lt;-na.omit(original_gene_list)
## sort the list in decreasing order 
gene_list = sort(gene_list, decreasing = TRUE)
## perform gseGO
gse_all&lt;- gseGO(geneList=gene_list, 
                 ont ="BP", 
                 keyType = "SYMBOL", 
                 nPerm = 1000, 
                 minGSSize = 3, 
                 maxGSSize = 800, 
                 pvalueCutoff = 1.00, 
                 verbose = TRUE, 
                 OrgDb = organism, 
                 pAdjustMethod = "none")


## count the gene number
gene_count&lt;- gse_all@result %&gt;% group_by(ID) %&gt;% summarise(count = sum(str_count(core_enrichment, "/")) + 1)

## merge with the original dataframe
dot_df&lt;- left_join(gse_all@result, gene_count, by = "ID") %&gt;% mutate(GeneRatio = count/setSize)

## Take 5 most significantly enriched terms
dot_df &lt;- dot_df[order(dot_df$pvalue),]
all_terms &lt;- dot_df$Description[c(1:5)]
dot_df_all &lt;- dot_df
dot_df_all$Celltype &lt;- "all"
</code></pre>
<h3 id="26-plot-gse-result-for-cell-based-deg">2.6. Plot GSE result for cell-based DEG</h3>
<pre><code>## bind cell type dataframes
namesX &lt;- unlist(names)
df &lt;- bind_rows(mget(namesX))

## bind celltype dataframe with all cells
df_total &lt;- rbind(df, dot_df_all)

## subset to only inlclude top GSE terms from all cells
df_total &lt;- subset(df_total, Description %in% all_terms)
length(unique(df_total$Description))

## [1] 5

## rename cell type names
df_total$Celltype[df_total$Celltype == "all"] &lt;- "All cells"
df_total$Celltype[df_total$Celltype == "Oligodendrocytes"] &lt;- "Oligoden."

## plot
gse_CellBased&lt;- ggplot(df_total, aes(x = GeneRatio, y=Description)) + 
  geom_point(aes(size = count, color = pvalue)) +
  theme_bw() +
  theme(strip.background = element_rect(colour=NA, fill=NA),
        strip.text = element_text(size = 12),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        axis.title.x = element_text(size = 12, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 12),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 12)) +
  scale_colour_gradient( low="red", limits=c(0.0,1)) +
  facet_wrap(~Celltype,  ncol=12) +
  ylab(NULL) +
  ggtitle("Cell-based: MAST")+
  scale_x_continuous(limits=c(0.1,1), labels = c("0.0","", "0.5","", "1.0"))
gse_CellBased
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/9ffb6fb9-6cec-40ba-9ded-740dbba20c13"> <br /> 
</p>

<p><strong>Figure 4. Cell-based DGE: Enrichment of the top 5 GO terms for GO-Biological Processes calculated for all cell types together across cell types.</strong>  DEGs with p-values &lt; 0.05 and log 2 fold-change &lt; -1 and &gt; 1 were used as the input for gene set enrichment analysis. The gene ratio, gene count, and p-value of the five terms in each cell type are shown.</p>
<hr />
<h2 id="3-sample-based-dge-enrichment-analysis">3. Sample-based DGE enrichment analysis</h2>
<h3 id="31-load-sample-based-all-cells-deg-file">3.1. load sample-based all cells DEG file</h3>
<pre><code>## load DEG csv file
all_sample &lt;- read.delim("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/pseudo_bulk_all_cells/PDvControl/DGE_AllCellsMainContrast HC vs PD.csv", header = T, sep = ",")
</code></pre>
<h3 id="32-load-sample-based-cell-type-groups-deg-files">3.2. load sample-based cell type groups DEG files</h3>
<pre><code>## list of cell types used for DEG contrasts
sample_based_cell_types &lt;- c('astro', 'DaN', 'endo', 'ependymal', 'excit', 'GABA', 'inhib', 'mg', 'Olig', 'opc', 'peri')

## create empty list object
list &lt;- list()

## load DEG csv files
for(i in unique(sample_based_cell_types)){
  list[paste0(i,"_sample")] &lt;- list(read.delim(paste0("/Users/mfiorini/Desktop/scRNA_pipeline/Manuscript/Manuscript_Figures/Smajic3/Step5/pseudo_bulk_celltype_groups/PDvControlbulk/info/DGE_",i,"MainContrast HC vs PD.csv"), header = T, sep = ","))
}

## create data frame for each cell type
list2env(list,envir=.GlobalEnv)
</code></pre>
<h3 id="33-clean-up-sample-based-deg-files">3.3. Clean up sample-based DEG files</h3>
<pre><code>## make list of cell type dataframe
CellType_df &lt;- list(astro_sample, DaN_sample, endo_sample, ependymal_sample, excit_sample, GABA_sample, inhib_sample, mg_sample, Olig_sample, opc_sample, peri_sample, all_sample)

## set names of list as cepp types
names(CellType_df) &lt;- c('astro_sample', 'DaN_sample', 'endo_sample', 'ependymal_sample', 'excit_sample', 'GABA_sample', 'inhib_sample', 'mg_sample', 'Olig_sample', 'opc_sample', 'peri_sample', "all_sample")

## list of columns that we want to keep
cols_keep &lt;- c("X","pvalue","log2FoldChange","padj")

## process each dataframe to only keep DEGs with a p-value &lt; 0.05 log2FC &gt; 1
list_SampleBased &lt;- list()

for (i in unique(names(CellType_df))){
  j &lt;- data.frame(CellType_df[i])
  colnames(j) &lt;- c('X', 'baseMean', 'log2FoldChange', 'lfcSE' ,'stat', 'pvalue', 'padj')
  j &lt;- j[,colnames(j) %in% cols_keep]
  j &lt;- subset(j, pvalue &lt;= 0.05)
  j &lt;- subset(j, log2FoldChange &gt;= 1 | log2FoldChange &lt;= -1)
  if (nrow(j &gt; 0)){
  j$cell_type &lt;- paste0(i)
  j$method &lt;- "sample_based"
  list_SampleBased[i] &lt;- list(j)
  }else{
    print(paste0("did not find any DEGs for ", i))
  }
  list_SampleBased[i]
}

## create data frame for each cell type
list2env(list_SampleBased,envir=.GlobalEnv)

## &lt;environment: R_GlobalEnv&gt;
</code></pre>
<h3 id="34-enrichr-cell-type-groups">3.4. EnrichR: Cell type groups</h3>
<pre><code>## do not run enrichr for all cell types
test &lt;- data.frame(names(list_SampleBased))
test &lt;- test %&gt;% 
  filter(!grepl('all', list_SampleBased))
cell_type_num &lt;- as.numeric(rownames(test))

## create empty list
SampleBased_gse_list &lt;- list()

## GSE loop
for (i in cell_type_num){
  ## create cell type specific dataframe
  j &lt;- data.frame(list_SampleBased[i])
  colnames(j) &lt;- c('X', 'log2FoldChange', 'pvalue', 'padj', 'cell_type', 'method')
  ## we want the log2 fold change 
  gene_list &lt;- j$log2FoldChange
  ## assign gene to log2fc
  names(gene_list) &lt;- j$X
  ## omit any NA values 
  gene_list&lt;-na.omit(gene_list)
  ## sort the list in decreasing order
  gene_list = sort(gene_list, decreasing = TRUE)
  ## perform gseGO
  gse_bulk&lt;- gseGO(geneList=gene_list, 
                            ont ="BP", 
                            keyType = "SYMBOL", 
                            nPerm = 1000, 
                            minGSSize = 3, 
                            maxGSSize = 800, 
                            pvalueCutoff = 1.00, 
                            verbose = TRUE, 
                            OrgDb = organism, 
                            pAdjustMethod = "none")
  ## organize results
  if (nrow(gse_bulk@result) &gt; 0){
  ## count the gene number
  gene_count&lt;- gse_bulk@result %&gt;% group_by(ID) %&gt;% summarise(count = sum(str_count(core_enrichment, "/")) + 1)
  ## merge with the original dataframe
  dot_df&lt;- left_join(gse_bulk@result, gene_count, by = "ID") %&gt;% mutate(GeneRatio = count/setSize)
  dot_df &lt;- dot_df[order(dot_df$pvalue),]
  dot_df$Celltype &lt;- names(list_SampleBased[i])
  ## store in list 
  SampleBased_gse_list[i] &lt;- list(dot_df)
  }else{
    print(paste0('no term enriched under specific pvalueCutoff for ',names(list_SampleBased[i])))
  }
}



## remove NULL (cell types with no enrichment results)
SampleBased_gse_list &lt;- SampleBased_gse_list[!sapply(SampleBased_gse_list,is.null)]

## set names of list element to cell types
names &lt;- list()
for (i in 1:length(SampleBased_gse_list)){
j &lt;- data.frame(SampleBased_gse_list[i])
names[i] &lt;- unique(j$Celltype)
}
names(SampleBased_gse_list) &lt;- names

## set to dataframe
list2env(SampleBased_gse_list,envir=.GlobalEnv)

## &lt;environment: R_GlobalEnv&gt;
</code></pre>
<h3 id="35-enrichr-all-cells">3.5. EnrichR: All cells</h3>
<pre><code>## retrieve logFC
original_gene_list &lt;- all_sample$log2FoldChange
## assign logFC to gene
names(original_gene_list) &lt;- all_sample$X
## omit any NA values 
gene_list&lt;-na.omit(original_gene_list)
## sort the list in decreasing order 
gene_list = sort(gene_list, decreasing = TRUE)
## perform gseGO
gse_all&lt;- gseGO(geneList=gene_list, 
                 ont ="BP", 
                 keyType = "SYMBOL", 
                 nPerm = 1000, 
                 minGSSize = 3, 
                 maxGSSize = 800, 
                 pvalueCutoff = 1.00, 
                 verbose = TRUE, 
                 OrgDb = organism, 
                 pAdjustMethod = "none")


## count the gene number
gene_count&lt;- gse_all@result %&gt;% group_by(ID) %&gt;% summarise(count = sum(str_count(core_enrichment, "/")) + 1)

## merge with the original dataframe
dot_df&lt;- left_join(gse_all@result, gene_count, by = "ID") %&gt;% mutate(GeneRatio = count/setSize)

## Take 5 most significantly enriched terms
dot_df &lt;- dot_df[order(dot_df$pvalue),]
all_terms &lt;- dot_df$Description[c(1:5)]
dot_df_all &lt;- dot_df
dot_df_all$Celltype &lt;- "all"
</code></pre>
<h3 id="36-plot-gse-result-for-cell-based-deg">3.6. Plot GSE result for cell-based DEG</h3>
<pre><code>## bind cell type dataframes
namesX &lt;- unlist(names)
df &lt;- bind_rows(mget(namesX))

## bind celltype dataframe with all cells
df_total &lt;- rbind(df, dot_df_all)

## subset to only inlclude top GSE terms from all cells
df_total &lt;- subset(df_total, Description %in% all_terms)
length(unique(df_total$Description))

## rename cell type names
df_total$Celltype[df_total$Celltype == "all"] &lt;- "All cells"
df_total$Celltype[df_total$Celltype == "DaN_sample"] &lt;- "DaN"
df_total$Celltype[df_total$Celltype == "endo_sample"] &lt;- "Endothelial"
df_total$Celltype[df_total$Celltype == "ependymal_sample"] &lt;- "Ependymal"
df_total$Celltype[df_total$Celltype == "excit_sample"] &lt;- "Excitatory"
df_total$Celltype[df_total$Celltype == "GABA_sample"] &lt;- "GABA"
df_total$Celltype[df_total$Celltype == "inhib_sample"] &lt;- "Inhibitory"
df_total$Celltype[df_total$Celltype == "mg_sample"] &lt;- "Microglia"
df_total$Celltype[df_total$Celltype == "Olig_sample"] &lt;- "Oligoden."
df_total$Celltype[df_total$Celltype == "opc_sample"] &lt;- "OPC"
df_total$Celltype[df_total$Celltype == "peri_sample"] &lt;- "Pericytes"
df_total$Celltype[df_total$Celltype == "astro_sample"] &lt;- "Astrocyte"

## plot
gse_CellBased&lt;- ggplot(df_total, aes(x = GeneRatio, y=Description)) + 
  geom_point(aes(size = count, color = pvalue)) +
  theme_bw() +
  theme(strip.background = element_rect(colour=NA, fill=NA),
        strip.text = element_text(size = 12),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        axis.title.x = element_text(size = 12, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 12),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 12)) +
  scale_colour_gradient( low="red", limits=c(0.0,1)) +
  facet_wrap(~Celltype,  ncol=12) +
  ylab(NULL) +
  ggtitle("Sample-based: DESeq2")+
  scale_x_continuous(limits=c(0.1,1), labels = c("0.0","", "0.5","", "1.0"))
gse_CellBased
</code></pre>
<p align="center">
<img src="https://github.com/neurobioinfo/scrnabox/assets/110110777/1e9126f3-cb17-4c80-8850-c5b457936370"> <br /> 
</p>

<p><strong>Figure 5. Sample-based DGE: Enrichment of the top 5 GO terms for GO-Biological Processes calculated for all cell types together across cell types.</strong>  DEGs with p-values &lt; 0.05 and log 2 fold-change &lt; -1 and &gt; 1 were used as the input for gene set enrichment analysis. The gene ratio, gene count, and p-value of the five terms in each cell type are shown.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../Dataset1/" class="btn btn-neutral float-left" title="Standard analysis track: Midbrain dataset"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../pbmc_download/" class="btn btn-neutral float-right" title="Downloading the PBMC dataset">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../Dataset1/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../pbmc_download/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
